---
title: "Supplementary Materials"
author: "Ben Nolan and  Nicholas Waters"
date: "9/18/2019"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Assembling candidate taxa from Extreme dataset
- command used for each of the organisms
- explanation of where some failed (which organisms didn't have usable references, which didn't have SRAs, etc)
- Where in the repo the resulting assembly is stored
## Generating Multiplee Sequence Alignments
- parameters used for mafft
- where in the repo the resulting alingmnts are

## Calcualting Alpha Diversity
```{r}

```
- copy the code to calculate the Shannon entropy here


## Assessing the provenance of the Silva strains
16s sequences in databases such a silva, greengenes,or RDP can come from several sources, namely:
- complete genomes (in Refseq, Genbank, etc)
- draft genomes (in the NCBI's Assembly database)
- 16s amplicons from Sanger sequencing
- 16s amplicons from high-throughput amplicon sequencing.

As repeated rDNA operons are difficult to assembly, Draft genome rDNA can (and often do contain a single rDNA).  This is problematic for species ideentification -- the rDNA recovered is not just one of /n/ rDNAs, but it can be a consensus "summary" rDNA resulting from trying to assemble the repeated region.   In organisms that have multiple rDNAs with high intrageno 
riboSeed has been show to generate high-quality reconstructions of each rDNA when benchmarked against hybrid assemblies. 

```{r}
ribo16s <- readLines("./results/databases/speciesdb/silva16/b_ovat_16.fa.mafft")
(ribo16s_headers <- ribo16s[grepl(">", ribo16s)])
```





