---
title: "Supplementary Materials"
author: "Ben Nolan and  Nicholas Waters"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  word_document: default
  html_document: default
editor_options:
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, cache.lazy = FALSE)
library(tidyverse); packageVersion("tidyverse")
library(diagram); packageVersion("diagram")
library(Biostrings); packageVersion("Biostrings")
#BiocManager::install("DECIPHER")
library(DECIPHER)
source("docs/shared.R")
```
##  Assessing SILVA's composition
```{bash, eval=FALSE}
if [ -f "~/Downloads/SILVA_132_SSURef_tax_silva.fasta.gz" ] ; then echo "already downloaded" ; else wget    https://www.arb-silva.de/fileadmin/silva_databases/release_132/Exports/SILVA_132_SSURef_tax_silva.fasta.gz   ; fi
```

First, we read in the silva db, and parse the headers for alter comparison. One thing we do it get the codes of what the prefixes from NCBI mean, and use that to determine with source of the sequences.
```{r parse_headers, cache=TRUE}
silva <- readRNAStringSet("~/Downloads/SILVA_132_SSURef_tax_silva.fasta.gz")

# codes_from_ncbi read in shared.R

silva_headers <- data.frame("V1"=names(silva), stringsAsFactors = F)
silva_headers$pre <- gsub("(\\D*).*", "\\1", silva_headers$V1)

#
silva_headers$id <- gsub("(.*?)\\.(.*)", "\\1", silva_headers$V1)
silva_headers$strain <- gsub("(.*?)\\.(.*?)\\.(.*)\\s.*;(.*)", "\\4", silva_headers$V1)
silva_headers$genus <- gsub("(.*?)\\s(.*?)\\s.*", "\\1", silva_headers$strain)
silva_headers$species <- gsub("(.*?)\\s(.*?)\\s.*", "\\2", silva_headers$strain)

#silva_headers$name <- gsub(">(.*?)\\s.*", "\\1", silva_headers$V1)
silva_headers$start <- as.numeric(gsub("(.*)\\.(.+)\\.(.*?)\\s.*", "\\2", silva_headers$V1))
silva_headers$stop <-  as.numeric(gsub("(.*)\\.(.+)\\.(.*?)\\s.*", "\\3", silva_headers$V1))
silva_headers$wgs <- nchar(silva_headers$pre) > 3 # see table

silva_headers$amplicon <- silva_headers$start < 10

genome_pres <- codes_from_ncbi %>% 
  filter(Type=="Genome project data") %>% 
  unnest(Prefix = strsplit(Prefix, ",")) %>% 
  select(Prefix) 
patent_pres <- codes_from_ncbi %>% filter(grepl("Patent", Type)) %>%
  unnest(Prefix = strsplit(Prefix, ",")) %>% 
  transform(Prefix=gsub(" ", "", Prefix)) %>% select(Prefix) 
silva_headers$cg <- silva_headers$pre %in% genome_pres$Prefix
silva_headers$patent <- silva_headers$pre %in% patent_pres$Prefix
silva_headers$htgs <- silva_headers$pre %in% c("AC", "DP")

table(silva_headers$cg) / nrow(silva_headers)
table(silva_headers$wgs) / nrow(silva_headers)
table(silva_headers$amplicon) / nrow(silva_headers)

silva_headers$category <- case_when(
  silva_headers$wgs ~ "Draft",
  silva_headers$htgs ~ "HTGS",
  silva_headers$patent ~ "Patent",
  silva_headers$cg ~ "Complete Genome",
  silva_headers$amplicon ~  "Amplicon")
round(table(silva_headers$category, useNA = "always") / nrow(silva_headers) * 100, 2)
silva_headers$matcher <- ifelse(silva_headers$cg, silva_headers$id,
                                ifelse(silva_headers$wgs, substr(silva_headers$id, start = 1, stop =( nchar(silva_headers$pre) + 2)), NA))
```

## Assembling candidate taxa from Extreme dataset
- command used for each of the organisms
- explanation of where some failed (which organisms didn't have usable references, which didn't have SRAs, etc)
- Where in the repo the resulting assembly is stored

27 strains from 26 species were in the extreme dataset from DADA2.

###### 17 strains have SRAs. Only 7 were completed with focusdb (two of which are 1 species. 6 species in total)

## Running FocusDB on the Extremes Dataset.
DADA2's Extremes dataset is comprised of the strains shown here in table 1. Note that we included Anaerocolumna xylanovorans, a more recent name for Clostridium xylanovorans.


## Running focusDB on the Extremes dataset

### Suppl Figure 1: Number of SRAs available for each species

A python script `sralist.py` was used to generate a list of the number of SRAs for a given species, based on parsing the results from sraFind.
```{bash eval=FALSE}
python scripts/sralist.py -s ../16db/sraFind-All-biosample-with-SRA-hits.tx -l ./docs/datasets/extreme_species_list.txt -o results/Extremes_sra_count
python scripts/sralist.py -s ../16db/sraFind-All-biosample-with-SRA-hits.txt -l ./docs/datasets/balanced_species_list.txt -o results/Balanced_sra_count
python scripts/sralist.py -s ../16db/sraFind-All-biosample-with-SRA-hits.txt -l ./docs/datasets/HMP_species_list.txt -o results/HMP_sra_count
```
```{r plot_sras}
srac <- data.frame("org"=NA, "count"=NA, stringsAsFactors = F)
srac <- srac[!is.na(srac),]
###
for (d in c("Extremes", "HMP", "Balanced")){
  thispath = file.path("results", paste0(d, "_sra_count"), "sralist.txt")
  tmp <- read.csv2(thispath, sep=":", col.names = c("org", "count"), stringsAsFactors = F, header=F)
  tmp$mock <- d
  srac <- rbind(srac, tmp)
}

srac$genus <- gsub("(.*) .*","\\1", srac$org)
srac$species <- gsub("(.*) (.*)","\\2", srac$org)
srac$has <- ifelse(srac$count == 0, "None Available", "SRAs Available")
(psrastrains <- ggplot(srac %>% filter(count > 0), aes(reorder(org,count), y=count, fill=genus, color=genus)) + 
    geom_bar(stat="identity") +
    coord_flip() + 
    scale_color_discrete(guide=F) +
    scale_fill_discrete(guide=F) +
    facet_wrap(~mock, scales = "free") +
    labs(title=paste0(nrow(srac %>% filter(count > 0)), " of ", nrow(srac), " strains have availabe SRA data "),
         fill="Genus",
         color="Genus",
         y= "Number of WGS SRAs", x="Strain") + 
    mytheme +
    theme(legend.position = c(.8, .3), 
          text = element_text(size=8))
)
ggsave("docs/figures/S1-sras.png", psrastrains, device = "png", dpi = 300, width = 8, height = 7)
```



###### Commands used for each of the 6 species with SRAs and reference genomes
```{bash eval=FALSE}
# executed on gruffalo at $SCRATCH/2019-09-19-16db/
# tried running with parallel, but in cases like this, its 
# not as efficient, as we wend up waiting for the organisms with more SRAs to finish
# cat extremes_orgs.txt | parallel -j 16 --colsep ' ' 16db  -o {1}_{2} -g {1}_{2}_genomes -n \""{1} {2}"\" --cores 2 --memory 80 --n_references 50 --n_SRAs 20
while read genus species; do 16db  -o ${genus}_${species} -g ${genus}_${species}_genomes -n "$genus $species" --cores 32 --memory 80 --n_references 50 --n_SRAs 20 --subassembler spades --threads 2  -v 1  ; done < extremes_orgs.txt
cat ./*/SUMMARY  > 2019-09-26-focusdb-results/summary_extremes_all
cp ./*/*ribo16s.fasta ./2019-09-26-focusdb-results/

```

### Figure 2: sra successes
```{r plot_success}

bal_data <- read.csv("./results/2019-10-07-focusdb-balanced/balanced_summary_all",
                     sep="\t", header=F, col.names = c("org", "status", "state", "messag"),
                     stringsAsFactors = F)
bal_data$mock <- "Balanced"
hmp_data <- read.csv("./results/2019-10-03-HMP-results/HMP_summary_all.tab",
                     sep="\t", header=F,
                     col.names = c("org", "status", "state", "messag"), 
                     stringsAsFactors = F)
hmp_data$mock <- "HMP"
ex_data <- read.csv("./results/2019-10-07-focusdb_extremes/Extremes_summary_all",
                    sep="\t", header=F,
                    col.names = c("org", "status", "state", "messag"), 
                    stringsAsFactors = F)
ex_data$mock <- "Extremes"
summary_data <- rbind(bal_data,ex_data, hmp_data)
globals <- summary_data %>% filter(state=="global") %>%
  select(mock, org, messag) %>% dplyr::rename(globalmessage=messag)
summary_data <- summary_data %>% filter(state!="global")
(globals$globalmessage <- ifelse(globals$globalmessage == "", "Success", globals$globalmessage))
# tidy things up;  this clarifies the immage a bit
(globals$globalmessage <- gsub("No 16s sequences detected in re-assemblies", "Success", globals$globalmessage))
# (globals$globalmessage <- gsub("No available.*", "No available references", globals$globalmessage))

p_summary_data_pre <- full_join(
  summary_data , globals, by=c("org", "mock")) %>% 
  select(mock, org, globalmessage, messag, state ) %>%  distinct()

p_summary_data <- right_join(srac,  p_summary_data_pre, by=c("org", "mock")) %>% 
  mutate(
    strain=org, 
    pmessage = case_when(
      startsWith(messag, "Unknown error") ~ "QC Fail",
      is.na(messag) ~ "QC Fail",
      startsWith(messag, "Library type Error") ~ "QC Fail",
      startsWith(messag, "Reads were") ~ "QC Fail",
      startsWith(messag, "Unable to process") ~ "QC Fail",
      messag=="" ~ "16S Recovered", 
      messag=="riboSeed unsuccessful" ~ "No 16S Recovered")) 

#  plot mock also in shared methods
mock_plots <- list()
for (grp in  c("HMP", "Extremes",  "Balanced")){
  thisgroup <- quo(grp)
  (mock_plots[[grp]] <- plot_mock(df=p_summary_data%>% filter(mock == (!!thisgroup))))
  ggsave(paste0("docs/figures/S2", grp, ".png"), mock_plots[[grp]], device = "png", dpi = 300, width = 12, height = 7)
}
###################
alpha_data <- p_summary_data %>%
  group_by(strain) %>% 
  mutate(alpha = ifelse(globalmessage=="Success", "gray10", "gray70")) %>% 
  select(strain, alpha) %>% distinct() %>% as.data.frame()
p_summary_data_B <- p_summary_data  %>%  arrange(pmessage) %>%
    group_by(strain) %>%
    mutate(thisid = row_number()) %>%
    transform(thisid = ifelse(is.na(messag), 0, thisid)) 
p_summary_data_B$strain <- factor(p_summary_data_B$strain, levels=rev(levels(factor(p_summary_data_B$strain))))

alpha_v <- alpha_data[order(alpha_data$strain, alpha_data$alpha), "alpha"]
(p_outcomesB <- ggplot(p_summary_data_B, 
                      aes(x=strain, y=thisid, shape=pmessage,
                          color=pmessage, fill=pmessage)) + 
    scale_shape_manual(guide = "legend", values = c(21, 22, 23,24)) + 
    geom_point(size=5)  + coord_flip() + mytheme +
    scale_colour_brewer(palette = "Set2") +
    scale_fill_brewer(palette = "Set2") +
    # unfortuantely facet wrapping throws off our beautiful alpha'd out names
    # due to those repeated
    #facet_wrap(~mock, nrow = 2, scales = "free") +
    labs("SRAs Proccessed by focusDB",
         y="Number of whole-genome sequencing SRAs", x="", color="Per SRA", fill="Per SRA", shape="Per SRA")  +
    theme(axis.text=element_text(size=13),
          axis.title.x =element_text(size=14),
          axis.text.y = element_text(colour = alpha_v))
)  
ggsave("docs/figures/2-success.png", p_outcomesB, device = "png", dpi = 300, width = 12, height = 7)

```

## Assessing uniqueness:
First, for simplicity, we read in 
```{r plot_uniqueness, cache=TRUE}
 
all_focus <- DNAStringSet() 
uniqueness <- data.frame("Organism"=NA, "Unique"=NA,  "Duplicated"=NA, "DB"=NA, stringsAsFactors = F)
uniqueness <- uniqueness[!is.na(uniqueness),]
# note that we only get the species level resultts here, later we address the genus level
counter <- 1
seq_paths <-  Sys.glob("./results/*/*_*_ribo16s.fasta"))
for (path in seq_paths){
  organism <- gsub("_", " ", gsub(".*\\/(.*)_ribo16s.fasta", "\\1", path))
  print(paste("processing", organism, counter, "of", length(seq_paths)))
  silva_subset <-silva[grepl(organism, names(silva))]
  nsilv <- length(silva_subset)
  nsilvdup = sum(duplicated(silva_subset), na.rm = TRUE)
  uniqueness <- rbind(
    uniqueness, 
    data.frame("Organism"=organism, "Unique"=nsilv-nsilvdup,  "Duplicated"=nsilvdup, "DB"="silv"))
  focus_subset <- readDNAStringSet(path)
  nfoc <- length(focus_subset)
  nfocdup = sum(duplicated(focus_subset), na.rm = TRUE)
  uniqueness <- rbind(
    uniqueness, 
    data.frame("Organism"=organism, "Unique"=nfoc-nfocdup,  "Duplicated"=nfocdup, "DB"="focus"))
  both_subsets  <- append(DNAStringSet(silva_subset), focus_subset)
  nboth <- length(both_subsets)
  nbothdup = sum(duplicated(both_subsets), na.rm = TRUE)
  uniqueness <- rbind(
    uniqueness, 
    data.frame("Organism"=organism, "Unique"=nboth-nbothdup,  "Duplicated"=nboth, "DB"="combined"))
  all_focus <- append(all_focus, both_subsets)
}
# set an arbtrary cutoff for popularity, just for plotting purposes
popthresh = 400; popular <- uniqueness %>% filter(DB=='silv') %>% filter(Unique > popthresh | Duplicated > popthresh)
p_uniqueness <- uniqueness %>% distinct() %>% gather(key="grp", value="val", -Organism, -DB ) %>%
  spread(key=DB, value=val) %>% 
  mutate(SILVA=combined-focus, focusDB=combined-silv) %>% 
  select(-silv, -combined, -focus)  %>%
  gather(key="DB", value="val", -Organism, -grp ) %>%
  # reverse levels for plottting
  mutate(Organism = factor(Organism), 
         Organism = factor(Organism, levels = rev(levels(Organism))))
(p_unique <- ggplot(
  p_uniqueness %>% mutate(fac = ifelse(Organism %in% popular$Organism, "Over-represented", "Under-represented")),
  aes(x=Organism, y=val, alpha=grp, group=grp, fill=DB)) +
    coord_flip() +
    facet_wrap(~fac, scales="free") +
    scale_fill_manual(values = viridis::viridis(n=3)[c(2, 1)]) +
    scale_alpha_manual(values = c(.2, 1)) + mytheme +
    labs(y="Number of Sequence", x="", alpha="", fill="Database") +
    theme(axis.text=element_text(size=13)) +
    geom_bar(stat="identity", position="stack") #+ facet_wrap(~grp) 
)
(p_unique_alt <- ggplot(
  p_uniqueness %>% mutate(fac = ifelse(Organism %in% popular$Organism, "Over-represented", "Under-represented"))%>% filter(grp=="Unique"),
  aes(x=Organism, y=val,  fill=DB)) +
    coord_flip() +
    facet_wrap(~fac, scales="free") +
    scale_fill_manual(values = viridis::viridis(n=3)[c(2, 1)]) +
    scale_alpha_manual(values = c(.2, 1)) + mytheme +
    labs(y="Number of Sequence", x="", alpha="", fill="Database") +
    theme(axis.text=element_text(size=13)) +
    geom_bar(stat="identity", position="stack") #+ facet_wrap(~grp) 
)
p_unique_perc <-  right_join(
  p_uniqueness %>% spread(grp, val) %>%
    dplyr::rename(org=Organism) %>%
    select(-Duplicated) %>% spread(DB, Unique) %>% 
    mutate(
      perc_inc=round((focusDB/SILVA) *100, 1 ),
      perc_of_tot=round(focusDB/((SILVA + focusDB)) *100, 1 ),
    ),
  srac %>% select(org, mock))%>% 
  filter(!is.na(perc_inc))
(p_unique_heat <- ggplot(p_unique_perc,
  aes(x=org, y=mock, fill=mock)) +
    coord_flip() +
    scale_fill_discrete(guide="none")+
    theme(rect = element_blank(),
          text=element_text(color="white"),
          axis.text.y  = element_blank(),
          axis.text.x  = element_text(size=10),
          line = element_blank(),
            
      )+
  geom_tile())
(p_uniqueV <- ggplot(p_unique_perc %>% select(-mock) %>% distinct(),
  aes(x=org, y=perc_inc)) +
    coord_flip() +
    scale_fill_manual(values = viridis::viridis(n=3)) +
    mytheme +
    labs(y="Percent increase in Unique Sequences", x="", alpha="", fill="Database") +
    theme(axis.text=element_text(size=13)) +
    geom_bar(stat="identity") #+ facet_wrap(~grp) 
)
(figure <- ggpubr::ggarrange(p_unique_heat, p_uniqueV, widths=c(1, 3)))
ggsave("docs/figures/3-unique.png", p_unique, device = "png", dpi = 300, width = 12, height = 7)
ggsave("docs/figures/3-unique_alt2.png", p_unique_alt, device = "png", dpi = 300, width = 12, height = 7)
ggsave("docs/figures/3-unique_alt.png", figure, device = "png", dpi = 300, width = 12, height = 7)

```


Here, we generate a table showing how many sequences (and unique seqeunces) are obtained from each level.  This is ued wiith table X in the paper
```{r tabledata}
A <- uniqueness %>% select(Organism, Unique, DB ) %>% distinct() %>%spread(key=DB, value=Unique)
B <- p_summary_data %>% select(strain, pmessage, state) %>% filter(!is.na(state)) %>% group_by(strain) %>% summarise(All_SRAs=n(), Good_SRAs=table(pmessage)["16S Recovered"]) 
right_join(B, A, by=c("strain"="Organism")) %>% 
  mutate(Genus=gsub("(.*) (.*)", "\\1", strain),
         Species=gsub("(.*) (.*)", "\\2", strain)) %>%
  select(Genus, Species, All_SRAs, Good_SRAs, focus, silv, combined)
 # %>% mutate(perc  = focus/silv*100) %>% select(perc) %>% summary()
```

## Generating Multiple Sequence Alignments
### Defining the regions of 16S in E coli
The coordinates for the common 16S primers correspond to E coli, so we selected this sequence.
Reference: 
 - https://www.ncbi.nlm.nih.gov/nuccore/NC_000913.3?report=fasta&from=4166659&to=4168200
 - the regions https://bmcbioinformatics.biomedcentral.com/track/pdf/10.1186/s12859-016-0992-
These are implemented in the `calculate-shannon-entropy` script.

```{bash eval=FALSE}
# Combining sequences
while read genus species; do if [ -f "./results/2019-09-26-focusdb-results/${genus}_${species}_ribo16s.fasta" ] ; then combine-focusdb-and-silva -d ~/Downloads/SILVA_132_SSURef_tax_silva.fasta -n "$genus $species" -S  ./results/2019-09-26-focusdb-results/${genus}_${species}_ribo16s.fasta --dna --lower  > ./results/extracted_sequences/${genus}_${species}_db.fasta ; fi; done < ./docs/extreme_species_list.txt
# alignment + trimming
while read genus species; do if [ -f "./results/2019-09-26-focusdb-results/${genus}_${species}_ribo16s.fasta" ] ; then align-and-trim-focusdb --i ./results/extracted_sequences/${genus}_${species}_db.fasta -o ./results/extracted_sequences/${genus}_${species}_db ; fi; done < ./docs/extreme_species_list.txt
#  
```


## Calcualting Alpha Diversity

```{r plot_shannon, eval=FALSE}
shannon <- read.csv(
  "./tmpalign.mafft.shannon", sep="\t", header = F, 
  col.names = c("pos", "region", "entropy", "entropy_subset"), 
  stringsAsFactors = F) %>% 
  gather(key="subset", value="entropy", -pos, -region) %>%
  mutate(grp = case_when(
    subset == "entropy" ~ "SILVA + focusDB",
    subset == "entropy_subset" ~ "SILVA"
  ))
str(shannon)

(p_diversity <- ggplot(shannon, aes(x=pos, color=grp, shape=grp, fill=region, y=entropy)) + 
    scale_shape_manual(values=c(21, 22)) + 
  geom_point(size=4, alpha=.5) + geom_smooth())
ggsave("docs/figures/4-diversity.png", p_diversity, device = "png", dpi = 300, width = 11, height = 7)
```


## Assessing the provenance of the Silva strains
16s sequences in databases such a silva, greengenes,or RDP can come from several sources, namely:
- complete genomes (in Refseq, Genbank, etc)
- draft genomes (in the NCBI's Assembly database)
- 16s amplicons from Sanger sequencing
- 16s amplicons from high-throughput amplicon sequencing.

As repeated rDNA operons are difficult to assembly, Draft genome rDNA can (and often do contain a single rDNA).  This is problematic for species ideentification -- the rDNA recovered is not just one of /n/ rDNAs, but it can be a consensus "summary" rDNA resulting from trying to assemble the repeated region.  

riboSeed has been show to generate high-quality reconstructions of each rDNA when benchmarked against hybrid assemblies. Here, we compare the 16s seqeunces from riboSeed reassembly of draft genomes to the initial (potentially collapsed) 16s seqeunces.

### Provenance of strains
First, we read in the silva headers for the strains. The header for the silva database constists of `>ACCESSION.start.end name`.  After a fruitless effort to rework sraFind's entrez calls to identify WGS master records with sraFind, I realized that info was already in pokaryotes.txt, and could be added as an afterthought.  As of release 0.8, that column is present as "WGS".  this allows us to relate the seqeunces in the SILVA headers to their original SRAs, so we can compare riboSeed's assemblies to those for which a complete genome is available, and assess the types of errors/collapses we see where only draft genomes are available. 

```{r eval=TRUE}
# read in the draft linked accessions and the complete genome linked accessions
# if (!file.exists("sraFind.tab")){
#   system("curl -O sraFind.tab -L https://raw.githubusercontent.com/nickp60/sraFind/master/sraFind.tab")
# }
# 
srafind <- read.csv("sraFind.tab", sep="\t", stringsAsFactors = F)
#  make a boolean mask for all the silva seqs we might care about
this_mask <- rep(FALSE, length(silva))
these_orgs <- srac[srac$count > 0, "org"] 
for (org in these_orgs){
  this_mask <- this_mask | grepl(org, names(silva))
}
silv_sub <- silva_headers[this_mask,]
# silv_sub_names <- data.frame(raw=names(silv_sub))
# silv_sub_names$pre <- gsub("(\\D*).*", "\\1", silv_sub_names$raw)
# silv_sub_names$id <- gsub("(.*?)\\.(.*)", "\\1", silv_sub_names$raw)
# silv_sub_names$matcher <- ifelse(substr(silv_sub_names$id, start = 1, 2) %in% unlist(genome_pres),
#                                  silv_sub_names$id, substr(silv_sub_names$id, start = 1, stop = length(silv_sub_names$pre) + 2))
# silv_sub_names$start <- gsub("(.*?)\\.(.*?)\\.(.*)\\s.*", "\\2", silv_sub_names$raw)
# silv_sub_names$end <- gsub("(.*?)\\.(.*?)\\.(.*?)\\s.*", "\\3", silv_sub_names$raw)
# silv_sub_names$strain <- gsub("(.*?)\\.(.*?)\\.(.*)\\s.*;(.*)", "\\4", silv_sub_names$raw)
# silv_sub_names$genus <- gsub("(.*?)\\s(.*?)\\s.*", "\\1", silv_sub_names$strain)
# silv_sub_names$species <- gsub("(.*?)\\s(.*?)\\s.*", "\\2", silv_sub_names$strain)
# 
# 
# Here, we constructt the column that is used to match with silva.  
# We take any NZ_ prefix off the complete genome recods.  If you know what this stands for and/or
# why silva doesnt use it, raise an issue on github, using the codeword "vermillion"
srafind$matcher <- ifelse(endsWith(srafind$WGS, "01"), srafind$WGS, 
                          gsub("NZ_", "", gsub("(.*?)\\.(.*)", "\\1", srafind$nuccore_first_chrom), fixed = T))
#View(srafind[grepl(".*_,*",  srafind$nuccore_first_chrom), ])
assembly_summary <- inner_join(
  silv_sub, 
  srafind %>% select(Assembly.Accession, matcher, platform,  run_SRAs) %>% filter(!is.na(run_SRAs)), 
  by="matcher")

# table(silv_sub$matcher %in% srafind$matcher)
# draft_silva_sras <- read.csv("./docs/accessions", sep="\t", header = F, 
#                              col.names = c("assembly", "silva", "Run", "SampleName", "ScientificName", "LibraryStrategy", "LibrarySource") ,
#                              stringsAsFactors = F)
# cg_silva_sras <- read.csv("./docs/cgaccessions", sep="\t", header = F, 
#                              col.names = c("assembly", "silva", "Run", "SampleName", "ScientificName", "LibraryStrategy", "LibrarySource") ,
#                              stringsAsFactors = F)
#                              
(p_summary_draft <- assembly_summary  %>% filter(wgs) %>% group_by(strain,  Assembly.Accession) %>% 
     select(strain, Assembly.Accession, id, start, stop) %>% distinct() %>%  summarize(n=n()))
(p_summary_cg    <-    assembly_summary  %>% filter(cg) %>% group_by(strain,  Assembly.Accession) %>% 
     select(strain, Assembly.Accession, id, start, stop) %>% distinct() %>%  summarize(n=n()))
p_summary_n16s <- assembly_summary %>% select(V1, strain, Assembly.Accession,category) %>%distinct() %>% group_by(strain,  Assembly.Accession) %>%  mutate(n=n()) %>%  select(-V1) %>% distinct()
(n16s <-  ggplot(p_summary_n16s, aes(x=n,  fill=category)) + geom_density(adjust=5, alpha=.6) + 
    labs(x="Number of 16S  in assembly", y="Relative Distriubtion", fill="") 
)
ggsave("docs/figures/S4-n16S.png", n16s, device = "png", dpi = 300, width = 6, height = 5)

table(p_summary_n16s[p_summary_n16s$category=="Complete Genome", "n"])
table(p_summary_n16s[p_summary_n16s$category!="Complete Genome", "n"])
# for example, 
# draft_fragilis <- p_summary_draft[grepl("fragilis", p_summary_draft$ScientificName),  ]

```


```{r}
build_aln_for_assembly <- function(db, assembly="GCA_000699725.1"){
  # get matching assembly from sraFind to get the complete genome/seqeunce accession
  this_silva <- unique(db[grepl(assembly, db$Assembly.Accession), "V1" ])
  if (length(this_silva) == 0) {
    warning("No matches for this assembly;  could be long reads?")
    return (NA)
  }
  # paste tofgether to make a query string with or's; this is because for assemblies, 
  # # different contigs will have different sequence names, but a single assembly accession
  this_silva_query <- gsub("(.*)\\|$", "\\1", paste0(unique(gsub("(.*?)\\..*", "\\1", this_silva)), collapse = "", sep="|"))
  # get matching SRAs for assembly, so we can use them to search the focusdb results
  this_focus <- unique(unlist(strsplit(db[grepl(assembly,  db$Assembly.Accession), "run_SRAs" ],",")))
  # build focusdb query string;  get rid of the *_\\d suffix we add to SRAs to keep themunique
  this_focus_query <- gsub("(.*)\\|$", "\\1", paste0(gsub("_\\d+", "",this_focus), collapse = "", sep="|"))
  
  # identify all seqeunces associated with this assembly accesssion
  silva_subset_seqs <- silva[grepl(this_silva_query,  names(silva))]
  # idenify the seqeunces from focusDB
  focus_subset_seqs <- unique(all_focus[grepl(this_focus_query, names(all_focus))])
  # merge, align the seqeunces
  # Silva by default is in RNA, so we type-cast to DNA
  # We add the silva seqs fist so we can check uniqueness by the presense of SRA seqs in the unique'd alignment
  aln <- DECIPHER::AlignSeqs(
    append(DNAStringSet(silva_subset_seqs), focus_subset_seqs)
  )
  uniques <- aln[!duplicated(subseq(aln, 5, -5))]
  
  # identity <- data.frame(name=names(focus_subset_seqs), stringsAsFactors = F, unique=F )
  # for (seq in identity$name){
  #   focus_subset_seqs[seq, "unique"] <- mapply(function(x, y ) sum(x!=y), strsplit(silva_subset_seqs,""), strsplit(focus_subset_seqs[seq],""))
  # }
  return (list("alignment"=aln, "unique_from_focusDB"=sum(names(focus_subset_seqs) %in% names(uniques)), "total"=length(aln)))
}

```

Next we  subsetted the dataframe to include only those that had sras from this study,(and by proxy, those that were seqeunced with illumina)
```{r}
only_illumina_assembly_summary <- assembly_summary # [assembly_summary$platform == "ILLUMINA", ]
only_illumina_assembly_summary$processed_in_this_study <- FALSE

for (sra in summary_data[summary_data$status=="PASS", "state"]){
  only_illumina_assembly_summary$processed_in_this_study <- grepl(
    sra, only_illumina_assembly_summary$run_SRAs,  fixed = T)  |  only_illumina_assembly_summary$processed_in_this_study
}
table(only_illumina_assembly_summary$processed_in_this_study )
only_illumina_assembly_summary <- only_illumina_assembly_summary%>% filter(processed_in_this_study)

# cellulosilyticus, excluded as no whole genomes short read available; reference was PacBio
GCF_001318345 <- build_aln_for_assembly(db=only_illumina_assembly_summary, assembly="GCA_001318345.1")
#BrowseSeqs(GCF_001318345[[1]])

#ovatus
GCF_001314995 <- build_aln_for_assembly(db=only_illumina_assembly_summary, assembly="GCA_001314995.1")
BrowseSeqs(GCF_001314995)
duplicated(subseq(GCF_001314995, 5, -5))

# thetaiotaomicron
GCF_001314975 <- build_aln_for_assembly(db=only_illumina_assembly_summary, assembly="GCA_001314975.1")
BrowseSeqs(GCF_001314975[[1]])
duplicated(subseq(GCF_001314975, 5, -5))

# vulgatus# also pac bio
GCF_001412315 <- build_aln_for_assembly(db=only_illumina_assembly_summary, assembly="GCA_001412315.1")
BrowseSeqs(subseq(GCF_001412315, 5, -5))
duplicated(subseq(GCF_001412315, 5, -5))

uniqueness_results = data.frame(assembly=NA, aln=NA, unique_focusdb=NA, total=NA, category=NA, stringsAsFactors = F)
uniqueness_results <-  uniqueness_results[!is.na(uniqueness_results), ]
nassemb <- nrow(only_illumina_assembly_summary)
for (i in 1:nassemb){
  print(paste(i, "of", nassemb ))
  assembly <- only_illumina_assembly_summary$Assembly.Accession[i]
  res <- build_aln_for_assembly(
    db=only_illumina_assembly_summary, 
    assembly=assembly
  )
  uniqueness_results <- rbind(
    uniqueness_results, 
    data.frame(assembly=assembly, aln=res[[1]], unique_focusdb=res[[2]], total=res[[3]], 
               category=only_illumina_assembly_summary$category[i])
  )
}
```

# assessing draft genome reassembly
Here,  looking at _B. ovatus_, we can see  that GCF_000699725.1 has 4 , so we dont need to reassemble those: rrndb shows that we can expect around  5 rDNAs for B ovatus.  For starters though, lets assess how we assembled the GCF_000699725.1.
The other three assemblies, we could re-assemble: GCF_000218325.1, GCF_000699665.1, and GCF_900107475.1 all only have a single rDNA.  Lets run the pipeline on the 4.

```{r}
ovatus = list()
draft_ovatus <- p_summary_draft[grepl("ovatus", p_summary_draft$ScientificName),  ]
for (i in 1:nrow(draft_ovatus)){
  try(ovatus[i] <- build_aln_for_assembly(db=draft_silva_sras, assembly=draft_ovatus$assembly[i]))
  outpath <- file.path("docs", "figures", "alignments", paste0(draft_ovatus$assembly[i], ".html"))
  try(BrowseSeqs(subseq(ovatus[[i]], 10, -10), highlight=0, threshold=0.5, htmlFile = outpath))
  
}
names(ovatus) <- draft_ovatus$assembly
for (s in ovatus)  try(print(duplicated(subseq(s, 10, -10))))


```

## Discussion
```{r calculate_needs_assembl}
sraFind_data <- read.table("../16db/sraFind-All-biosample-with-SRA-hits.txt", header=T, sep="\t", stringsAsFactors = F)
tbassembled <- sraFind_data %>% filter (nuccore_first_chrom=="") %>% filter(!is.na(study_acc)) %>% nrow()
tbreleased  <- sraFind_data %>% filter (nuccore_first_chrom=="") %>% filter(is.na(study_acc)) %>% nrow()
```

## Execution time


```{r}
library(lubridate)
alog <- readLines("./results/2019-10-03-hmp-focusdb/combined_log.txt")
alog <- alog[1:length(alog)]
lo <- data.frame(dt=gsub("(.*?),.* - .*? - (.*)","\\1", alog), stringsAsFactors = F)
lo$message  <-  gsub("(.*?) - .*? - (.*)","\\2", alog)
lo$datetime  <- ymd_hms(lo$dt)
# caused by broken lines in log
lo <- lo[!is.na(lo$datetime), ]
lo <- lo[grepl("Processing|Found SRAs|Downloading genomes|checking reference|Downloading|Obtaining average|Finding best|Quality trimming reads|Downsampling reads|Running riboSeed|Sum of|Skipping riboSeed", lo$message), ]
lo$bug <- ifelse(grepl("Processing \\D", lo$message),  gsub("Processing ","", lo$message), NA)
lo$sra <- ifelse(grepl("Downloading .*\\d+", lo$message),  gsub("Downloading ","", lo$message), ifelse(
  	grepl("Processing", lo$message), "GLOBAL", NA))
lo <- lo %>% fill(sra) %>% fill(bug) %>% group_by(sra, bug) %>% mutate(keep = !any(grepl("Skipping riboSeed", message))) %>% filter(keep) %>% select(-keep)
# fix download by removing reference to SRA
lo$message <- gsub("Downloading .*\\d+","Downloading SRA", gsub("(\\D+?)\\s(\\D+?)[$ ].*","\\1 \\2", lo$message))
# fix riboseed messgage by removing number
lo$message <- gsub("Processing \\D+", "Processing org", lo$message)
# fix riboseed messgage by removing number
lo$message <- gsub("\\d*", "", lo$message)
# get rid of missing
lof <- lo %>%  group_by(bug) %>% mutate(remove = any(grepl("Quality trimming|Obtaining", message )& sra == "GLOBAL")) %>% filter(!remove) %>% select(-remove) %>% filter(bug != "Actinomyces odontolyticus")
lot <- lof %>% select(-dt) %>% filter(sra != "SRR1819776") %>% spread(key=message, value=datetime)
lot$org_done = lot$`Downloading SRA`[1]
for (i in 1:nrow(lot)){
  if (i !=nrow(lot)){
    if (lot$sra[i] != "GLOBAL"){
      if (lot$sra[i+1]  !=  "GLOBAL"){
        # if next item is another SRR
        lot$org_done[i] <- lot$`Downloading SRA`[i+1]
      }else {
        # If next item is a global message
        lot$org_done[i] <- lot$`Processing  riboSeed runs`[i+1]
      } 
    }
  }
}
# per SRA steps
lot$Download_SRA <- difftime(lot$`Quality trimming` ,  lot$`Downloading SRA`, units="m")
lot$Trim <- difftime(lot$`Downsampling reads`, lot$`Quality trimming`, units="m")
#lot$Downsample<- lot$org_done  - lot$`Downsampling reads`
# global steps
lot$Assemble <- difftime(lot$`Sum of`, lot$`Processing  riboSeed runs`, units = "m")
lot$Download_refs <- difftime(lot$`checking reference`, lot$`Downloading genomes`, units="m")
lot$Find_SRAs <- difftime(lot$`Found SRAs:`, lot$`Processing org`, units="m")

tlot  <- lot %>% select(bug, sra, Download_SRA, Trim, Assemble, Download_refs, Find_SRAs) %>%
  gather(key="step", value="s", -bug, -sra) %>%
  filter(!is.na(s)) %>%
  mutate(stage = ifelse(grepl("ad_SRA|Trim|Downsample", step), "SRA", "Global"))


(p_timing <- ggplot(tlot, aes(x=step, y=s, color= bug)) + 
    geom_boxplot(color="black", fill=NA) +
    geom_jitter(width = .2) + 
    scale_y_continuous() +
    scale_color_discrete(guide="none") +
    coord_flip() + 
    labs(title="Time per stage", x="Step", y="Minutes") +
    facet_wrap(~stage, scales="free"))  

ggsave("docs/figures/S3-time.png", p_timing, device = "png", 
       dpi = 300, width = 11, height = 7)

```