---
title: "Supplementary Materials"
author: "Ben Nolan and  Nicholas Waters"
date: "9/18/2019"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
```

## Assembling candidate taxa from Extreme dataset
- command used for each of the organisms
- explanation of where some failed (which organisms didn't have usable references, which didn't have SRAs, etc)
- Where in the repo the resulting assembly is stored

27 strains from 26 species were in the extreme dataset from DADA2.

###### 17 strains have SRAs. Only 7 were completed with focusdb (two of which are 1 species. 6 species in total)

## Running FocusDB on the Extremes Dataset.
DADA2's Extremes dataset is comprised of the strains shown here in table 1. Note that we included Anaerocolumna xylanovorans, a more recent name for Clostridium xylanovorans.

```{r}
res <- read.csv("./docs/extreme_species_list.txt", sep="\t", col.names = "strains")
```
```{}
Bacteroides cellulosilyticus DSM 14838: 3
Bacteroides fragilis ATCC 23745: 112
Bacteroides massiliensis JCM 12982: 1
Bacteroides ovatus DSM 1896: 12
Bacteroides thetaiotaomicron DSM 2079: 14
Bacteroides uniformis DSM 6597: 15
Bacteroides vulgatus DSM 1447: 14
Clostridium xylanovorans DSM 12503: 1: Renamed to Anaerocolumna xylanovorans
Coprococcus comes ATCC 27758: 4
Parabacteroides distasonis JCM 13400: 18
Parabacteroides distasonis JCM 13401: 18
Parabacteroides merdae DSM 19495: 2
Parabacteroides sp. D13 BEI HM-77: 3
Paraprevotella clara DSM 19731: 1
Prevotella copri DSM 18205: 1
Roseburia intestinalis DSM 14610: 1
Roseburia inulinivorans DSM 16841: 3
```


###### No SRAs
```{}
Bacteroides eggerthii BEI HM-210: 0
Barnesiella intestinihominis DSM 21032: 0
Clostridium celatum JCM 1394: 0
Clostridium cocleatum DSM 1551: 0 
Clostridium methylpentosum DSM 5476: 0
Clostridium phytofermentans ATCC 700394: 0 
Eubacterium rectale DSM 17629: 0
Howardella ureilytica DSM 15118: 0
Prevotella buccalis ATCC 35310: 0 
Ruminococcus gnavus ATCC 29149: 0
```

## Running focusDB on the Extremes dataset
```
# contents of extremes_orgs.txt manifests
Bacteroides cellulosilyticus
Bacteroides fragilis
Bacteroides massiliensis
Bacteroides ovatus
Bacteroides thetaiotaomicron
Bacteroides uniformis
Bacteroides vulgatus
Clostridium xylanovorans
Anaerocolumna xylanovorans
Coprococcus comes
Parabacteroides distasonis
Parabacteroides merdae
Parabacteroides sp. 
Paraprevotella clara
Prevotella copri 
Roseburia intestinalis
Roseburia inulinivorans
Bacteroides eggerthii 
Barnesiella intestinihominis
Clostridium celatum 
Clostridium cocleatum 
Clostridium methylpentosum
Clostridium phytofermentans
Eubacterium rectale 
Howardella ureilytica 
Prevotella buccalis 
Ruminococcus gnavus 
```

###### Number of SRAs available for each species

Simple python script that for a species in a list, returns the number of occurances in the sraFind file.
```{bash eval=FALSE}
python scripts/sralist.py -s ~/Desktop/16db/py16db/sraFind-All-biosample-with-SRA-hits.txt -l ~/Desktop/focusdb/docs/extreme_species_list.txt -o results/extreme_sra_count
```

###### table for 6 successful strains
```{r}
table <- read.csv("./docs/successful_strains_table.txt", header=TRUE, sep = " ")
knitr::kable(table)
```

###### Commands used for each of the 6 species with SRAs and reference genomes
```{bash eval=FALSE}
# executed on gruffalo at $SCRATCH/2019-09-19-16db/
# cat extremes_orgs.txt | parallel -j 16 --colsep ' ' 16db  -o {1}_{2} -g {1}_{2}_genomes -n \""{1} {2}"\" --cores 2 --memory 80 --n_references 50 --n_SRAs 20
cat ./*/SUMMARY  > 2019-09-22-focusdb-results/results_summary.tab

16db -o /HDD1/Ben/16db/results/b_cellulosilyticus -n "Bacteroides cellulosilyticus" -g /HDD1/Ben/16db/results/b_cellulosilyticus/genomes/ -S 15 --memory 100 --cores 12

16db -o /HDD1/Ben/16db/results/b_fragilis -n "Bacteroides fragilis" -g /HDD1/Ben/16db/results/b_fragilis/genomes/ -S 15 --memory 100 --cores 12

16db -o /HDD1/Ben/16db/results/b_ovatus -n "Bacteroides ovatus" -g /HDD1/Ben/16db/results/b_ovatus/genomes/ -S 15 --memory 100 --cores 12

16db -o /HDD1/Ben/16db/results/b_thetaiotaomicron -n "Bacteroides thetaiotaomicron" -g /HDD1/Ben/16db/results/b_thetaiotaomicron/genomes/ -S 15 --memory 100 --cores 12

16db -o /HDD1/Ben/16db/results/b_vulgatus -n "Bacteroides vulgatus" -g /HDD1/Ben/16db/results/b_vulgatus/genomes/ -S 15 --memory 100 --cores 12

16db -o /HDD1/Ben/16db/results/p_distasonis -n "Parabacteroids distasonis" -g /HDD1/Ben/16db/results/p_distasonis/genomes/ -S 15 --memory 100 --cores 12
```


## Generating Multiplee Sequence Alignments
- parameters used for mafft
- where in the repo the resulting alingmnts are

## Calcualting Alpha Diversity
```{r}

```
- copy the code to calculate the Shannon entropy here


## Assessing the provenance of the Silva strains
16s sequences in databases such a silva, greengenes,or RDP can come from several sources, namely:
- complete genomes (in Refseq, Genbank, etc)
- draft genomes (in the NCBI's Assembly database)
- 16s amplicons from Sanger sequencing
- 16s amplicons from high-throughput amplicon sequencing.

As repeated rDNA operons are difficult to assembly, Draft genome rDNA can (and often do contain a single rDNA).  This is problematic for species ideentification -- the rDNA recovered is not just one of /n/ rDNAs, but it can be a consensus "summary" rDNA resulting from trying to assemble the repeated region.  

riboSeed has been show to generate high-quality reconstructions of each rDNA when benchmarked against hybrid assemblies. Here, we compare the 16s seqeunces from riboSeed reassembly of draft genomes to the initial (potentially collapsed) 16s seqeunces.

### Provenance of b_ovatus strains
First, we read in the silva headers for the b_ovatus strains. The header for the silva database constists of `>ACCESSION.start.end name`.  We are not concerned with the complete genomes, so we will try to match the acessions with the 
```{r}
library(data.table)
ribo16s <- readLines("./results/databases/speciesdb/silva16/b_ovat_16.fa.mafft")
sraFind_data <- fread("https://raw.githubusercontent.com/nickp60/sraFind/master/results/sraFind-All-biosample-with-SRA-hits.txt")
(ribo16s_headers <- ribo16s[grepl(">", ribo16s)])
#  get the assembly
#  esearch -db nuccore -query JNHG01000060 | elink -target assembly | esummary |xtract -pattern DocumentSummary -element AssemblyAccession
#  get the SRA
cmd_file <-  file.path("docs/tmp_link_accessions")
file.remove(cmd_file)
file.remove(file.path("docs/b_ovatus_sras"))
for (header in ribo16s_headers){
  # if not from riboSeed
  header <- gsub(">", "", header)
  if (!grepl("extracted", header)){
    id = gsub("(.*?)\\.(.*)", "\\1", header )
    cmd = paste0("OLDIFS=\"$IFS\"; assembly=`esearch -db nuccore -query ",id, " | elink -target biosample | elink -target assembly  | esummary | xtract -pattern DocumentSummary -element AssemblyAccession`; IFS=$\"\n\"; for line in `esearch -db nuccore -query ", id, " | elink -target biosample | elink -target sra |  esummary -format runinfo -mode xml | xtract -pattern Row -element Run -element SampleName -element ScientificName -element LibraryStrategy -element LibrarySource`; do echo -e \"${assembly}\t",   strsplit(header, split = " ")[[1]][1], "\t$line\"; done; IFS=$OLDIFS")
    write(cmd, file=cmd_file, append=TRUE)
  }
}

   
```
Then we can run those commands (from the terminal, doesn't tend to work from Rstudio):

```{bash, eval=FALSE}
# note this doesn't work from Rstudio due to how Rstudio get the system path.
bash tmp_link_accessions > b_ovatus_sras
```

Now, we have a list any available SRAs for the list of strains in the Silva database, with columns for the name/region (from silva), the assembly, the sample name, scientific name, tech, type, and SRA.  Note that the silva name referes to the sequence, not the assembly, so thats why we need  to fetch the column for the aassembly as well.

Now, we can look at how many rDNAs are represented in the database.
```{r}
names <- c("assembly", "silva", "Run", "SampleName", "ScientificName", "LibraryStrategy", "LibrarySource")
ovatus_sras <- read.csv("./docs/b_ovatus_sras", sep="\t", header = F, col.names = names)
(ovatus_summary <- ovatus_sras  %>% group_by(assembly) %>% 
    select(assembly, silva) %>% distinct() %>%  summarize(n=n()))
left_join(ovatus_summary%>%filter(n!=1), ovatus_sras) %>% select(Run) %>% distinct()
left_join(ovatus_summary%>%filter(n==1), ovatus_sras) %>% select(Run)
```
Here, we can see  that GCF_000699725.1 and GCF_001314995.1 have 4 and 5 rDNAs in Silva, so we dont need to reassemble those: rrndb shows that we can expect around  5 rDNAs for B ovatus.

The other three assemblies, we could re-assemble: GCF_000218325.1, GCF_000699665.1, and GCF_900107475.1 all only have a single rDNA.  Lets run the pipeline on the 4 

```
16db --SRAs SRR9067260 SRR1006208 SRR1006209 SRR4236982 -o ovatus_draft -g ovatus_genomes -n "Bacteroides ovatus" --cores 4 --memory 8 
16db --SRAs SRR1006214  SRR1006215 SRR2637631 SRR2637689 SRR2641383 -o ovatus_complete -g ovatus_genomes -n "Bacteroides ovatus" --cores 4 --memory 8
```

Now we can compare.

