---
title: "Supplementary Materials"
author: "Ben Nolan and  Nicholas Waters"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  word_document: default
  html_document: default
editor_options:
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, cache.lazy = FALSE)
library(tidyverse); packageVersion("tidyverse")
library(diagram); packageVersion("diagram")
library(Biostrings); packageVersion("Biostrings")
#BiocManager::install("DECIPHER")
library(DECIPHER)

mytheme <- ggplot2::theme_minimal() + 
  ggplot2::theme(
    rect = element_rect(fill = "transparent"),
    #plot.background = element_rect(fill = "#FAFAFA", color=NA),
    plot.background = element_rect(fill = "transparent", color=NA),
    axis.text = element_text(size=8),
    strip.background = element_rect(fill = "grey95", color=NA),
    strip.text = element_text(face="bold"),
    axis.title  = element_text(size=10),
    panel.grid.minor.x = element_blank(),
    title = element_text(size=14),
    plot.caption=element_text(hjust = 0),
    legend.text =  element_text(size=12), 
    plot.subtitle = element_text(size=12, colour = "grey60"),
    #legend.position = "bottom"
  )
```
##  Assessing SILVA's composition
```{bash, eval=FALSE}
#cat ~/Downloads/SILVA_132_SSURef_tax_silva.fasta.gz |gunzip  | grep ">" > ./silva132ref_headers.txt
```

```{r parse_headers, cache=TRUE}
silva <- readRNAStringSet("SILVA_132_SSURef_tax_silva.fasta")

codes_from_ncbi <- read.csv(text="Prefix	Database	Type
BA,DF,DG,LD	DDBJ	CON division
AN	EMBL	CON division
CH,CM,DS,EM, EN,EP,EQ,FA, GG,GL,JH,KB, KD,KE,KI,KK, KL,KN,KQ,KV, KZ,ML	NCBI	CON division
C,AT,AU,AV,BB, BJ,BP,BW,BY,CI, CJ,DA,DB,DC, DK,FS,FY,HX, HY,LU	DDBJ	EST
F	EMBL	EST
H,N,T,R,W,AA,AI, AW,BE,BF,BG, BI,BM,BQ,BU, CA,CB,CD,CF, CK,CN,CO,CV, CX,DN,DR,DT, DV,DW,DY,EB, EC,EE,EG,EH, EL,ES,EV,EW, EX,EY,FC,FD, FE,FF,FG,FK, FL,GD,GE,GH, GO,GR,GT,GW, HO,HS,JG,JK, JZ	GenBank	EST
D,AB,LC	DDBJ	Direct submissions
V,X,Y,Z,AJ,AM, FM,FN,HE,HF, HG,FO,LK,LL, LM,LN,LO,LR, LS,LT,OA,OB, OC,OD,OE	EMBL	Direct submissions
U,AF,AY,DQ,EF, EU,FJ,GQ,GU, HM,HQ,JF,JN, JQ,JX,KC,KF, KJ,KM,KP,KR, KT,KU,KX,KY, MF,MG,MH,MK, MN	GenBank	Direct submissions
AP	DDBJ	Genome project data
BS	DDBJ	Chimpanzee genome data
AL,BX,CR,CT, CU,FP,FQ,FR	EMBL	Genome project data
AE,CP,CY	GenBank	Genome project data
AG,DE,DH,FT, GA,LB	DDBJ	GSS
B,AQ,AZ,BH,BZ, CC,CE,CG,CL, CW,CZ,DU,DX, ED,EI,EJ,EK, ER,ET,FH,FI, GS,HN,HR,JJ, JM,JS,JY,KG, KO,KS,MJ	GenBank	GSS
AK	DDBJ	cDNA projects
AC,DP	GenBank	HTGS
E,BD,DD,DI,DJ, DL,DM,FU,FV, FW,FZ,GB,HV, HW,HZ,LF,LG, LV,LX,LY,LZ, MA,MB,MC	DDBJ	Patents
A,AX,CQ,CS,FB, GM,GN,HA,HB, HC,HD,HH,HI, JA,JB,JC,JD, JE,LP,LQ,MP, MQ,MR,MS	EMBL	Patents (nucleotide only)
I,AR,DZ,EA,GC, GP,GV,GX,GY, GZ,HJ,HK,HL, KH,MI,MM,MO	GenBank	Patents (nucleotide)
G,BV,GF	GenBank	STS
BR	DDBJ	TPA
BN	EMBL	TPA
BK	GenBank	TPA
HT,HU	DDBJ	TPA CON division
BL,GJ,GK	GenBank	TPA CON division
EZ,HP,JI,JL, JO,JP,JR,JT, JU,JV,JW,KA	GenBank	TSA
FX,LA,LE,LH, LI,LJ	DDBJ	TSA
S	GenBank	From journal scanning
AD	GenBank	From GSDB
AH	GenBank	Segmented set header
AS	GenBank	Other - not currently being used
BC	GenBank	MGC project
BT	GenBank	FLI-cDNA projects
J,K,L,M	GenBank	from GSDB direct submissions
N	GenBank and DDBJ	N0-N2 were used intially by both groups but have been removed from circulation, N2-N9 are ESTs
AAAA-AZZZ, JAAA-JZZZ, LAAA-LZZZ, MAAA-MZZZ, NAAA-NZZZ, PAAA-PZZZ, QAAA-QZZZ, RAAA-RZZZ, SAAA-SZZZ, VAAA-VZZZ	GenBank	WGS
AAAAAA-AZZZZZ	GenBank	WGS
BAAA-BZZZ	DDBJ	WGS
BAAAAA-BZZZZZ	DDBJ	WGS
CAAA-CZZZ, FAAA-FZZZ, OAAA-OZZZ, UAAA-UZZZ	EMBL	WGS
CAAAAA-CZZZZZ	EMBL	WGS
DAAA-DZZZ	GenBank	WGS TPA
DAAAAA-DZZZZZ	GenBank	WGS TPA
EAAA-EZZZ	DDBJ	WGS TPA
GAAA-GZZZ	GenBank	TSA
HAAA-HZZZ	EMBL	TSA
IAAA-IZZZ	DDBJ	TSA
TAAA-TZZZ	DDBJ	Targeted Gene Projects
KAAA-KZZZ	GenBank	Targeted Gene Projects
AAAAA-AZZZZ	DDBJ	MGA", sep="\t", stringsAsFactors=F)

silva_headers <- data.frame("V1"=names(silva), stringsAsFactors = F)
silva_headers$pre <- gsub(">(\\D*).*", "\\1", silva_headers$V1)
#silva_headers$name <- gsub(">(.*?)\\s.*", "\\1", silva_headers$V1)
silva_headers$start <- as.numeric(gsub("(.*)\\.(.+)\\.(.*?)\\s.*", "\\2", silva_headers$V1))
silva_headers$stop <-  as.numeric(gsub("(.*)\\.(.+)\\.(.*?)\\s.*", "\\3", silva_headers$V1))
silva_headers$wgs <- nchar(silva_headers$pre) > 3 # see table

silva_headers$amplicon <- silva_headers$start < 10
genome_pres <- codes_from_ncbi %>% 
  filter(Type=="Genome project data") %>% 
  unnest(Prefix = strsplit(Prefix, ",")) %>% 
  select(Prefix) 
patent_pres <- codes_from_ncbi %>% filter(grepl("Patent", Type)) %>%
  unnest(Prefix = strsplit(Prefix, ",")) %>% 
  transform(Prefix=gsub(" ", "", Prefix)) %>% select(Prefix) 
silva_headers$cg <- silva_headers$pre %in% genome_pres$Prefix
silva_headers$patent <- silva_headers$pre %in% patent_pres$Prefix
silva_headers$htgs <- silva_headers$pre %in% c("AC", "DP")

table(silva_headers$cg) / nrow(silva_headers)
table(silva_headers$wgs) / nrow(silva_headers)
table(silva_headers$amplicon) / nrow(silva_headers)

silva_headers$category <- case_when(
  silva_headers$wgs ~ "Draft",
  silva_headers$htgs ~ "HTGS",
  silva_headers$patent ~ "Patent",
  silva_headers$cg ~ "Complete Genome",
  silva_headers$amplicon ~  "Amplicon")
round(table(silva_headers$category, useNA = "always") / nrow(silva_headers) * 100, 2)
```


## Generating the flowchart
```{r make_flowchart}
png("docs/figures/1-flowchart.png", width = 6, height = 6, res = 300, units = "in")
par(mar = c(1, 1, 1, 1))
openplotmat() 
cexd <- .5
boxsize = 0.065
thisrady = 0.04
thisradx = 0.14
cols <- c(.2,.35,  .5, .65, .8) # 5 cols
rows <- seq(.1, .9, .11) - .05  # 6 

shadowbox(box.type = "rect", mid = c(0.48, 0.52), box.col="gray95",lcol="gray85",
          rady = 0.25, radx = 0.3, angle = 0, shadow.size=0)
shadowbox(box.type = "rect", mid = c(0.49, 0.51),  box.col="gray95", lcol="gray85",
          rady = 0.25, radx = 0.3, angle = 0, shadow.size=0)
shadowbox(box.type = "rect", mid = c(0.5, 0.50), box.col="gray95", lcol="gray85",
          rady = 0.25, radx = 0.3, angle = 0, shadow.size=0)
# shadowbox(box.type = "rect", mid = c(0.52, 0.52),
#           rady = 0.4, radx = 0.2, angle = 0)
straightarrow(from =  c(cols[1], rows[8]),
              to = c(cols[3], rows[7]),
              arr.length=.2, arr.width=.2,
              lwd = 1, arr.pos = 0.6)
straightarrow(from =  c(cols[5], rows[8]),
              to = c(cols[3], rows[7]),
              arr.length=.2, arr.width=.2,
              lwd = 1, arr.pos = 0.65)

textellipse(mid=c(cols[1], 0.95), 
        lab = "RefSeq Database",
         col = "black",rady=thisrady, radx=thisradx, angle=0,
         shadow.col = "#7b6c60", shadow.size = 0.005, cex = cexd) 
textellipse(c(cols[5], .95),
         boxsize, lab = "SRA Database",
         box.col = "#ffe4ca",rady=thisrady, radx=thisradx, 
         shadow.col = "#7b6c60", shadow.size = 0.005, cex = cexd)
textrect(c(cols[1], rows[8]), 
         boxsize, lab = "1. Download References\n(plentyofbugs)",
         box.col = "#ffe4ca",rady=thisrady, radx=thisradx,
         shadow.col = "#7b6c60", shadow.size = 0.005, cex = cexd) 
textrect(c(cols[5], rows[8]), boxsize, lab = "2. Download SRAs\n(sraFind, sra-tools)",
         box.col = "#ffe4ca",rady=thisrady, radx=thisradx, 
         shadow.col = "#7b6c60", shadow.size = 0.005, cex = cexd) 
textrect(c(cols[3], rows[7]), boxsize, lab = "For each SRA:",
         box.col = "white",rady=thisrady, radx=thisradx, 
         shadow.col = "grey80", shadow.size = 0.005, cex = cexd) 
##  Sra Stuff
textrect(c(cols[2], rows[6]), boxsize, lab = "3. Check Read Length;\nQuality Trim (sickle)",
         box.col = "white",rady=thisrady, radx=thisradx, 
         shadow.col = "grey80", shadow.size = 0.005, cex = cexd) 
textrect(c(cols[2], rows[5]), boxsize, lab = "4. Downsample if \nCoverage > 50x (seqtk)",
         box.col = "white",rady=thisrady, radx=thisradx, 
         shadow.col = "grey80", shadow.size = 0.005, cex = cexd) 
##  Reference Selection
# textrect(c(cols[4], rows[4]), boxsize, lab = "Check Reference \nGenome Completeness",
#          box.col = "white",rady=thisrady, radx=thisradx, 
#          shadow.col = "grey80", shadow.size = 0.005, cex = cexd) 
textrect(c(cols[4], .55), boxsize, lab = "5. Select Best Reference\n(plentyofbugs)",
         box.col = "white",rady=thisrady, radx=thisradx, 
         shadow.col = "grey80", shadow.size = 0.005, cex = cexd) 
textrect(c(cols[3], rows[4]), boxsize, lab = "6. De fere novo Assembly\n(riboSeed)",
         box.col = "white",rady=thisrady, radx=thisradx, 
         shadow.col = "grey80", shadow.size = 0.005, cex = cexd) 
textrect(c(cols[3], rows[3]), boxsize, 
            lab = "7. Extract 16s from \n new assemblies (barrnap)",
            box.col = "#a0ff9e",rady=thisrady, radx=thisradx,
             shadow.col = "#83d283", shadow.size = 0.005, cex = cexd) 
## Finalize
textrect(c(cols[2], rows[2]), boxsize,  lab = "(8*) Extract Species's 16S\nfrom focusDB and Silva",
         box.col = "#ffe4ca",rady=thisrady, radx=thisradx, 
         shadow.col = "#7b6c60", shadow.size = 0.005, cex = cexd) 
textrect(c(cols[4], rows[2]), boxsize, 
            lab = "(9*) Align and Trim\n(mafft, TrimAl)",
            box.col = "#d5b8af",rady=thisrady, radx=thisradx,
             shadow.col = NA, shadow.size = 0.005, cex = cexd) 
straightarrow(from =  c(cols[3], rows[1]),
              to = c(cols[3], rows[1] - .05),
              arr.length=.2, arr.width=.2,
              lwd = 1, arr.pos = 0.65)

textrect(c(cols[3], rows[1]), boxsize, 
            lab = "Use with DADA2, etc",
            box.col = "white",rady=thisrady, radx=thisradx,
             shadow.col = NA, shadow.size = 0.005, cex = cexd) 
dev.off()

```

## Assembling candidate taxa from Extreme dataset
- command used for each of the organisms
- explanation of where some failed (which organisms didn't have usable references, which didn't have SRAs, etc)
- Where in the repo the resulting assembly is stored

27 strains from 26 species were in the extreme dataset from DADA2.

###### 17 strains have SRAs. Only 7 were completed with focusdb (two of which are 1 species. 6 species in total)

## Running FocusDB on the Extremes Dataset.
DADA2's Extremes dataset is comprised of the strains shown here in table 1. Note that we included Anaerocolumna xylanovorans, a more recent name for Clostridium xylanovorans.

```{r}
res <- read.csv("./docs/extreme_species_list.txt", sep="\t", col.names = "strains")
```



## Running focusDB on the Extremes dataset

### Suppl Figure 1: Number of SRAs available for each species

A python script `sralist.py` was used to generate a list of the number of SRAs for a given species, based on parsing the results from sraFind.
```{bash eval=FALSE}
python scripts/sralist.py -s ~/Desktop/16db/py16db/sraFind-All-biosample-with-SRA-hits.txt -l ~/Desktop/focusdb/docs/extreme_species_list.txt -o results/extreme_sra_count
```
```{r plot_sras}
srac <- read.csv2("results/extreme_sra_count/sralist.txt", sep=":", col.names = c("strain", "count"), stringsAsFactors = F, header=F)

srac$genus <- gsub("(.*) .*","\\1", srac$strain)
srac$species <- gsub("(.*) (.*)","\\2", srac$strain)
srac$has <- ifelse(srac$count == 0, "None Available", "SRAs Available")
(psrastrains <- ggplot(srac %>% filter(count > 0), aes(reorder(strain,count), y=count, fill=genus, color=genus)) + 
  geom_bar(stat="identity") +
  coord_flip() + 
  labs(title=paste0(nrow(srac %>% filter(count > 0)), " of ", nrow(srac), " strains have availabe SRA data "),
       fill="Genus",
       color="Genus",
       y= "Number of WGS SRAs", x="Strain") + 
    mytheme +
   theme(legend.position = c(.8, .3), 
         text = element_text(size=8))
  )
ggsave("docs/figures/S1-sras.png", psrastrains, device = "png", dpi = 300, width = 6, height = 5)
```



###### Commands used for each of the 6 species with SRAs and reference genomes
```{bash eval=FALSE}
# executed on gruffalo at $SCRATCH/2019-09-19-16db/
# tried running with parallel, but in cases like this, its 
# not as efficient, as we wend up waiting for the organisms with more SRAs to finish
# cat extremes_orgs.txt | parallel -j 16 --colsep ' ' 16db  -o {1}_{2} -g {1}_{2}_genomes -n \""{1} {2}"\" --cores 2 --memory 80 --n_references 50 --n_SRAs 20
while read genus species; do 16db  -o ${genus}_${species} -g ${genus}_${species}_genomes -n "$genus $species" --cores 32 --memory 80 --n_references 50 --n_SRAs 20 --subassembler spades --threads 2  -v 1  ; done < extremes_orgs.txt
cat ./*/SUMMARY  > 2019-09-26-focusdb-results/summary_all
cp ./*/*ribo16s.fasta ./2019-09-26-focusdb-results/

```

### Figure 2: sra successes
```{r plot_success}
summary_data <- read.csv("results/2019-09-26-focusdb-results/summary_all", sep="\t", header=F,
                         col.names = c("org", "status", "state", "messag"), stringsAsFactors = F)
globals <- summary_data %>% filter(state=="global") %>%
  select(org, messag) %>% dplyr::rename(globalmessage=messag)
(globals$globalmessage <- ifelse(globals$globalmessage == "", "Success", globals$globalmessage))
# tidy things up
(globals$globalmessage <- gsub("No 16s sequences detected in re-assemblies", "No reassembled 16S", globals$globalmessage))
# (globals$globalmessage <- gsub("No available.*", "No available references", globals$globalmessage))


p_summary_data_pre <- merge(
  summary_data %>% filter(state!="global"),
  globals, by="org", all=T) %>% 
  select(org, globalmessage, messag, state ) %>%  distinct()

p_summary_data <- merge(srac,  p_summary_data_pre, by.x="strain", by.y="org") %>% 
  mutate(
    pmessage = case_when(
      is.na(messag) ~ "QC Fail",
      startsWith(messag, "Reads were") ~ "QC Fail",
      startsWith(messag, "Unable to process") ~ "QC Fail",
      messag=="" ~ "16S Recovered", 
      messag=="riboSeed unsuccessful" ~ "No 16S Recovered")) %>%
  arrange(pmessage) %>%
  group_by(strain) %>%
  mutate(thisid = row_number()) %>%
 transform(thisid = ifelse(is.na(messag), 0, thisid)) 
#p_summary_data %>% group_by(strain) %>% summarize(n=n()) %>% View() 
# put in original number of orrganisms, for plotting purposes()
(p_outcomes <- ggplot(p_summary_data, 
                      aes(x=strain, y=thisid, shape=pmessage,
                          color=globalmessage, fill=globalmessage)) + 
    scale_shape_manual(guide = "legend", values = c(21, 22, 23,24)) + 
    geom_point(size=5)  + coord_flip() + mytheme +
    labs("SRAs Proccessed by focusDB",
         y="Number of SRAs", x="", color="Run Status", fill="Run Status", shape="SRA status")  +
    theme(axis.text=element_text(size=13))
)  
ggsave("docs/figures/2-success.png", p_outcomes, device = "png", dpi = 300, width = 11, height = 7)

```

## Assessing uniqueness:

```{r plot_uniqueness, cache=TRUE}
all_focus <- DNAStringSet()
uniqueness <- data.frame("Organism"=NA, "Unique"=NA,  "Duplicated"=NA, "DB"=NA, stringsAsFactors = F)
uniqueness <- uniqueness[!is.na(uniqueness),]
for (organism in c("Bacteroides cellulosilyticus", "Bacteroides fragilis", "Bacteroides ovatus",
                   "Bacteroides thetaiotaomicron", "Bacteroides vulgatus", 
                    "Bacteroides uniformis", "Parabacteroides distasonis")){
  print(paste("processing", organism))
  silva_subset <-silva[grepl(organism, names(silva))]
  thistab <- table(duplicated(silva_subset))
  uniqueness <- rbind(
    uniqueness, 
    data.frame("Organism"=organism, "Unique"=thistab[[1]],  "Duplicated"=thistab[[2]], "DB"="silv"))
  focus_subset <- readDNAStringSet(paste0("./results/extracted_sequences/", gsub(" ",  "_", organism), "_db.fasta"))
  thistab <- table(duplicated(focus_subset))
  uniqueness <- rbind(
    uniqueness, 
    data.frame("Organism"=organism, "Unique"=thistab[[1]],  "Duplicated"=thistab[[2]], "DB"="combined"))
  focus_subset_only <- focus_subset[grepl("extracted_seed",names(focus_subset))]
  all_focus  <- append(all_focus, focus_subset_only)
  Fa <- table(duplicated(focus_subset_only))[1]
  Tr <- table(duplicated(focus_subset_only))[2]
  if (is.na(Tr)) Tr <- 0
  uniqueness <- rbind(
    uniqueness, 
    data.frame("Organism"=organism, "Unique"=Fa,  "Duplicated"=Tr, "DB"="focus"))
}
p_uniqueness <- uniqueness %>%  gather(key="grp", value="val", -Organism, -DB ) %>%
  spread(key=DB, value=val) %>% 
  mutate(SILVA=combined-focus, focusDB=combined-silv) %>% 
  select(-silv, -combined, -focus)  %>%
  gather(key="DB", value="val", -Organism, -grp ) %>%
  # reverse levels for plottting
  mutate(Organism = factor(Organism), 
                    Organism = factor(Organism, levels = rev(levels(Organism))))
(p_unique <- ggplot(
  p_uniqueness,
  aes(x=Organism, y=val, alpha=grp, group=grp, fill=DB)) +
    coord_flip() +
    scale_fill_manual(values = viridis::viridis(n=3)[c(2, 1)]) +
    scale_alpha_manual(values = c(.2, 1)) + mytheme +
    labs(y="Number of Sequence", x="", alpha="", fill="Database") +
    theme(axis.text=element_text(size=13)) +
    geom_bar(stat="identity", position="stack") #+ facet_wrap(~grp) 
)

ggsave("docs/figures/3-unique.png", p_unique, device = "png", dpi = 300, width = 9, height = 5)

```

```{r tabledata}
A <- uniqueness %>% select(Organism, Unique, DB ) %>%spread(key=DB, value=Unique)
B <- p_summary_data %>% select(strain, pmessage, state) %>% filter(!is.na(state)) %>% group_by(strain) %>% summarise(All_SRAs=n(), Good_SRAs=table(pmessage)["16S Recovered"]) 
right_join(B, A, by=c("strain"="Organism")) %>% 
  mutate(Genus=gsub("(.*) (.*)", "\\1", strain),
         Species=gsub("(.*) (.*)", "\\2", strain)) %>%
  select(Genus, Species, All_SRAs, Good_SRAs, focus, silv, combined)
```

## Generating Multiple Sequence Alignments
### Defining the regions of 16S in E coli
The coordinates for the common 16S primers correspond to E coli, so we selected this sequence.
Reference: 
 - https://www.ncbi.nlm.nih.gov/nuccore/NC_000913.3?report=fasta&from=4166659&to=4168200
 - the regions https://bmcbioinformatics.biomedcentral.com/track/pdf/10.1186/s12859-016-0992-
These are implemented in the `calculate-shannon-entropy` script.

```{bash eval=FALSE}
# Combining sequences
while read genus species; do if [ -f "./results/2019-09-26-focusdb-results/${genus}_${species}_ribo16s.fasta" ] ; then combine-focusdb-and-silva -d ~/Downloads/SILVA_132_SSURef_tax_silva.fasta -n "$genus $species" -S  ./results/2019-09-26-focusdb-results/${genus}_${species}_ribo16s.fasta --dna --lower  > ./results/extracted_sequences/${genus}_${species}_db.fasta ; fi; done < ./docs/extreme_species_list.txt
# alignment + trimming
while read genus species; do if [ -f "./results/2019-09-26-focusdb-results/${genus}_${species}_ribo16s.fasta" ] ; then align-and-trim-focusdb --i ./results/extracted_sequences/${genus}_${species}_db.fasta -o ./results/extracted_sequences/${genus}_${species}_db ; fi; done < ./docs/extreme_species_list.txt
#  
```


## Calcualting Alpha Diversity

```{r plot_shannon, eval=FALSE}
shannon <- read.csv(
  "./tmpalign.mafft.shannon", sep="\t", header = F, 
  col.names = c("pos", "region", "entropy", "entropy_subset"), 
  stringsAsFactors = F) %>% 
  gather(key="subset", value="entropy", -pos, -region) %>%
  mutate(grp = case_when(
    subset == "entropy" ~ "SILVA + focusDB",
    subset == "entropy_subset" ~ "SILVA"
  ))
str(shannon)

(p_diversity <- ggplot(shannon, aes(x=pos, color=grp, shape=grp, fill=region, y=entropy)) + 
    scale_shape_manual(values=c(21, 22)) + 
  geom_point(size=4, alpha=.5) + geom_smooth())
ggsave("docs/figures/4-diversity.png", p_diversity, device = "png", dpi = 300, width = 11, height = 7)
```


## Assessing the provenance of the Silva strains
16s sequences in databases such a silva, greengenes,or RDP can come from several sources, namely:
- complete genomes (in Refseq, Genbank, etc)
- draft genomes (in the NCBI's Assembly database)
- 16s amplicons from Sanger sequencing
- 16s amplicons from high-throughput amplicon sequencing.

As repeated rDNA operons are difficult to assembly, Draft genome rDNA can (and often do contain a single rDNA).  This is problematic for species ideentification -- the rDNA recovered is not just one of /n/ rDNAs, but it can be a consensus "summary" rDNA resulting from trying to assemble the repeated region.  

riboSeed has been show to generate high-quality reconstructions of each rDNA when benchmarked against hybrid assemblies. Here, we compare the 16s seqeunces from riboSeed reassembly of draft genomes to the initial (potentially collapsed) 16s seqeunces.

### Provenance of b_ovatus strains
First, we read in the silva headers for the b_ovatus strains. The header for the silva database constists of `>ACCESSION.start.end name`.  We are not concerned with the complete genomes, so we will try to match the acessions with the 
```{r, eval=FALSE}
# ribo16s <- readLines("./results/databases/speciesdb/silva16/b_ovat_16.fa.mafft")
# (ribo16s_headers <- ribo16s[grepl(">", ribo16s)])
#  get the assembly
#  esearch -db nuccore -query JNHG01000060 | elink -target assembly | esummary |xtract -pattern DocumentSummary -element AssemblyAccession
#  get the SRA
get_links_for_accessions  <- function(cmd_file,  res_file, level="cg"){
  try(file.remove(cmd_file))
  #file.remove(res_file)
  counter <- 0
  for (organism in c("Bacteroides cellulosilyticus", "Bacteroides fragilis", "Bacteroides ovatus",
                     "Bacteroides thetaiotaomicron", "Bacteroides vulgatus", 
                     "Bacteroides uniformis", "Parabacteroides distasonis")){
    focus_subset <- readDNAStringSet(paste0("./results/extracted_sequences/", gsub(" ",  "_", organism), "_db.fasta"))
    for (header in names(focus_subset)){
      # if not from riboSeed
      if (!grepl("extracted", header)){
        pre <- gsub("(\\D*).*", "\\1", header)
        # if compeltte genome:
         pre %in% genome_pres$Prefix
        # if a wgs record, see above
        iscg <-  pre %in% genome_pres$Prefix
        iswgs <-  nchar(gsub("(\\D*).*", "\\1", header)) > 3
        # skip if not wgs
        if (! isdraft & level=="draft") {next}
        if (! iscg & level=="cg") {next}
        id = gsub("(.*?)\\.(.*)", "\\1", header )
        if (counter == 0){ {write("OLDIFS=\"$IFS\"; IFS=$\"\n\"", file=cmd_file, append=TRUE) }}
        cmd = paste0("assembly=`esearch -db nuccore -query ",id, " | elink -target biosample | elink -target assembly  | esummary | xtract -pattern DocumentSummary -element AssemblyAccession`; for line in `esearch -db nuccore -query ", id, " | elink -target biosample | elink -target sra |  esummary -format runinfo -mode xml | xtract -pattern Row -element Run -element SampleName -element ScientificName -element LibraryStrategy -element LibrarySource`; do echo -e \"${assembly}\t",   strsplit(header, split = " ")[[1]][1], "\t$line\"; done")
        write(cmd, file=cmd_file, append=TRUE)
      }
    }
  }
  write("IFS=$OLDIFS", file=cmd_file, append=TRUE)
}

#  for draft genomes  
cmd_file <-  file.path("docs/tmp_link_accessions")
res_file <-  file.path("docs/accessions")
get_links_for_accessions(cmd_file,  res_file, level="draft")
# for complete genomes sequencing
cmd_file <-  file.path("docs/cg_link_accessions")
res_file <-  file.path("docs/cgaccessions")
get_links_for_accessions(cmd_file,  res_file, level="cg")

```

Then we can run those commands (from the terminal, doesn't tend to work from Rstudio):

```{bash, eval=FALSE}
# note this doesn't work from Rstudio due to how Rstudio get the system path.
bash docs/tmp_link_accessions > docs/accessions
bash docs/cg_link_accessions > docs/cgaccessions
```

Now, we have a list any available SRAs for the list of strains in the Silva database, with columns for the name/region (from silva), the assembly, the sample name, scientific name, tech, type, and SRA.  Note that the silva name refers to the sequence, not the assembly, so thats why we need  to fetch the column for the aassembly as well.

Now, we can look at how many rDNAs are represented in the database.
```{r eval=TRUE}
# read in the draft linked accessions and the complete genome linked accessions
names <- c("assembly", "silva", "Run", "SampleName", "ScientificName", "LibraryStrategy", "LibrarySource")
draft_silva_sras <- read.csv("./docs/accessions", sep="\t", header = F, col.names = names,stringsAsFactors = F)
cg_silva_sras <- read.csv("./docs/cgaccessions", sep="\t", header = F, col.names = names,stringsAsFactors = F)

(p_summary_draft <- draft_silva_sras  %>% group_by(ScientificName,  assembly) %>% 
    select(ScientificName, assembly, silva) %>% distinct() %>%  summarize(n=n()))
(p_summary_cg    <-    cg_silva_sras  %>% group_by(ScientificName,  assembly) %>% 
    select(ScientificName, assembly, silva) %>% distinct() %>%  summarize(n=n()))


# for example, 
draft_fragilis <- p_summary_draft[grepl("fragilis", p_summary_draft$ScientificName),  ]
```


```{r}
build_aln_for_assembly <- function(db, assembly="GCF_000699725.1"){
  # get matching assembly from entrez link results to get the complete genome/seqeunce accession
  this_silva <- unique(db[grepl(assembly, db$assembly), "silva" ])
  # paste tofgether to make a query string with or's; this is because for assemblies, 
  # # different contigs will have different sequence names, but a single assembly accession
  this_silva_query <- gsub("(.*)\\|$", "\\1", paste0(gsub("(.*?)\\..*", "\\1", this_silva), collapse = "", sep="|"))
  # get matching SRAs for assembly, so we can use them to search the focusdb results
  this_focus <- unique(db[grepl(assembly, db$assembly), "Run" ])
  # build focusdb query string;  get rid of the *_\\d suffix we add to SRAs to keep themunique
  this_focus_query <- gsub("(.*)\\|$", "\\1", paste0(gsub("_\\d+", "",this_focus), collapse = "", sep="|"))
  
  # identify all seqeunces associated with this assembly accesssion
  silva_subset_seqs <- silva[grepl(this_silva_query,  names(silva))]
  # idenify the seqeunces from focusDB
  focus_subset_seqs <- all_focus[grepl(this_focus_query, names(all_focus))]
  # merge, align the seqeunces
  # Silva by default is in RNA, so we type-cast to DNA
  aln <- DECIPHER::AlignSeqs(
    append(focus_subset_seqs, DNAStringSet(silva_subset_seqs))
  )
  return (aln)
}

```

```{r}
# cellulosilyticus, excluded as no whole genomes short read available; reference was PacBio
GCF_001318345 <- build_aln_for_assembly(db=cg_silva_sras, assembly="GCF_001318345.1")
BrowseSeqs(GCF_001318345)
#ovatus
GCF_001314995 <- build_aln_for_assembly(db=cg_silva_sras, assembly="GCF_001314995.1")
BrowseSeqs(subseq(GCF_001314995, 5, -5))
duplicated(subseq(GCF_001314995, 5, -5))
# thetaiotaomicron
GCF_001314975 <- build_aln_for_assembly(db=cg_silva_sras, assembly="GCF_001314975.1")
BrowseSeqs(subseq(GCF_001314975, 5, -5))
duplicated(subseq(GCF_001314975, 5, -5))

# vulgatus# also pac bio
GCF_001412315 <- build_aln_for_assembly(db=cg_silva_sras, assembly="GCF_001412315.1")
BrowseSeqs(subseq(GCF_001412315, 5, -5))
duplicated(subseq(GCF_001412315, 5, -5))

```

# assessing draft genome reassembly
Here,  looking at _B. ovatus_, we can see  that GCF_000699725.1 has 4 , so we dont need to reassemble those: rrndb shows that we can expect around  5 rDNAs for B ovatus.  For starters though, lets assess how we assembled the GCF_000699725.1.
The other three assemblies, we could re-assemble: GCF_000218325.1, GCF_000699665.1, and GCF_900107475.1 all only have a single rDNA.  Lets run the pipeline on the 4.

```{r}
ovatus = list()
draft_ovatus <- p_summary_draft[grepl("ovatus", p_summary_draft$ScientificName),  ]
for (i in 1:nrow(draft_ovatus)){
  try(ovatus[i] <- build_aln_for_assembly(db=draft_silva_sras, assembly=draft_ovatus$assembly[i]))
  outpath <- file.path("docs", "figures", "alignments", paste0(draft_ovatus$assembly[i], ".html"))
  try(BrowseSeqs(subseq(ovatus[[i]], 10, -10), highlight=0, threshold=0.5, htmlFile = outpath))
  
}
names(ovatus) <- draft_ovatus$assembly
for (s in ovatus)  try(print(duplicated(subseq(s, 10, -10))))


```

## Discussion
```{r calculate_needs_assembl}
sraFind_data <- read.table("../16db/sraFind-All-biosample-with-SRA-hits.txt", header=T, sep="\t", stringsAsFactors = F)
tbassembled <- sraFind_data %>% filter (nuccore_first_chrom=="") %>% filter(!is.na(study_acc)) %>% nrow()
tbreleased  <- sraFind_data %>% filter (nuccore_first_chrom=="") %>% filter(is.na(study_acc)) %>% nrow()
```