---
title: 'Mining underutilized whole-genome sequencing projects to improve 16S rRNA databases'
date: "`r format(Sys.time(), '%d %B, %Y')`"
toc: false
fontsize: 10pt
papersize: a4
geometry: margin=2cm
header-includes:
     - \usepackage{setspace}
     - \doublespacing
     - \usepackage{lineno}
     - \linenumbers
     - \usepackage{float}
     - \title{\fontseries{m}\selectfont Mining underutilized whole-genome sequencing projects to improve 16S rRNA databases}
     - \author{\protect\parbox{\textwidth}{\protect\centering\small Ben Nolan,$^{1,2}$ Florence Abram,$^{1}$ Fiona Brennan,$^{1,3}$ Ashleigh Holmes,$^{4}$  Vincent O’Flaherty$^{1}$,\newline Leighton Pritchard$^{2,5\ast}$ and Nicholas R Waters$^{1,3}$ \newline \newline \small{\textit {$^{1}$Functional Environmental Microbiology, School of Natural Sciences, National University of Ireland, Galway, Ireland}} $^{2}$Microbial Ecology Laboratory, Microbiology, School of Natural Sciences, National University of Ireland, Galway, Ireland \newline \small{\textit {$^{2,5}$Information and Computational Sciences, James Hutton Institute, Invergowrie, Dundee DD2 5DA, Scotland}} \newline \small{\textit {$^{3}$Soil and Environmental Microbiology, Environmental Research Centre, Teagasc, Johnstown Castle, Wexford, Ireland}}\newline \small{\textit {$^{4}$Cell and Molecular Sciences, James Hutton Institute, Invergowrie, Dundee DD2 5DA, Scotland}}\newline  \small{\textit {$^{5}$Strathclyde Institute of Pharmacy and Biomedical Sciences, University of Strathclyde, Glasgow, G1 1XQ, Scotland}}\newline \footnotesize{$^\ast$To whom correspondence should be addressed$:$ nickp60@gmail.com or leighton.pritchard@strath.ac.uk \newline \newline}}}
     - \usepackage{caption}
     - \captionsetup{font={stretch=1.0}}
     - \captionsetup[table]{font={stretch=1.0}}
output:
  bookdown::pdf_document2: 
     fig_caption: true
     keep_tex: true
     #latex_engine: xelatex
editor_options:
  chunk_output_type: console
bibliography: focusdb_references.bib
abstract: Current approaches to interpreting 16S rDNA amplicon data are hampered by several factors. Among these are database inaccuracy or incompleteness, sequencing inaccuracy, and biased DNA/RNA extraction. Existing 16S rDNA databases source the majority of sequences from deposited amplicon sequences, draft genomes, and complete genomes. Most of the draft genomes available are assembled from short reads. However, repeated ribosomal regions are notoriously difficult to assemble well from short reads, and as a consequence the short-read-assembled 16S rDNA region may be an amalgamation of different loci within the genome. This complicates high-resolution community analysis, as draft genome's 16S sequence may be a chimera of multiple loci; in such cases, it may not be an exact match to the appropriate reference sequence. We present Focus16, a pipeline for improving 16S rDNA databases by mining NCBI's Sequence Read Archive for underexploited whole-genome sequencing runs. Using riboSeed (a genome assembly tool for correcting rDNA misassembly), Focus16 provides a way to augmment 16S rDNA databases with high-quality re-assembled sequences. In this study, we show an average increase of 60% in the number of unique sequences for given species compared to the 16S rRNA database alone. We augmented the widely-used SILVA 16S database with the novel sequences disclosed by Focus16 and re-processed sequences from the different datasets with DADA2; using this augmented SILVA database increased the number of ASVs that could be assigned taxonomic annotations. Further, fine-scale classification improved by revealing ambiguities; for example, an ASV may have been assigned to be a particular genus, but the additional sequences in the augmented database revealed that the ASV may belong to two or more different genera.  We demonstrate that improvements can be made to taxonomic classification by incorporating these re-assembled 16S rRNA sequences, and we invite the community to expand our work to augment existing 16S rDNA databases such as SILVA, GreenGenes, and RDP.
---
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE)
knitr::opts_chunk$set(cache=FALSE)
knitr::opts_chunk$set(fig.pos = 'H')
library(tidyverse)
library(diagram)

library(kableExtra)
# draft 4 but I think we have a name
toolname <- "Focus16"
```


# Introduction
The use of genetic markers for microbial classification has seen explosive growth over the past decade   [@liu2012popular;@boers_understanding_2019]. The 16S rRNA gene is the standard utilised in the assessment of prokaryotic community composition by amplicon sequencing [@fukuda2016molecular]. 16S rRNA has been used for community analysis in diverse environments such as the gut microbiota of cattle and pigs  [@avila2018prediction], soil [@santamaria2018soil], marine environments   [@dang2000bacterial], and the gut [@jovel_characterization_2016]. The success of this method hinges on the presence of the 16S rRNA gene in all domains and its slow evolution; thus the rDNA regions can be targeted with primers but the amplicon sequences maintain enough diversity that organisms can be differentiated at the genus or species level [@woese_phylogenetic_1977;@woo2008then]. 


Microbial genomes have a range of 16S rRNA gene copy numbers (GCNs), from the many Mycobacterum with a single copy to  *Photobacterium damselae* Phdp Wu-121 with 21 copies [@vetrovsky_variability_2013;@stoddard_rrndb_2015;@acinas2004divergence]. There may be variability between each 16S rDNA copy [@sun2013intragenomic]; this can negatively impact 16S rRNA classification in two ways.  First, in taxa with low variability, diversity estimates by overestimating taxa with higher GCN and underestimating those with low GCN; for instance, several sequences from isolates of the same species  would be difficult to differentiate from the copies within a single species. Second, some organisms have sufficient sequence variability between copies that they may be assigned different taxonomic classifications; indeed,  certain extremophiles have been reported to possess very high 16S rRNA copy heterogeneity, up to 9.3% sequence variation in some species  [@sun2013intragenomic]; this is well beyond the 97% or 99% clustering thresholds commonly used for community analysis.  Although these repeats complicate community analysis, each instance of the 16S rRNA contains valuable information. An ideal analysis framework would utilize a database incorporating this information to both correct for copy number variation between organisms in a community, and correctly relate 16S rRNA variants to each organism. 


Amongst the most widely used 16S rRNA databases for bacteria and archaea are Greengenes [@desantis2006greengenes], SILVA [@quast2012silva], and the Ribosomal Database Project (RDP) [@cole2005ribosomal]. Each contains 16S rRNA sequences derived from multiple major international nucleotide sequence databases, principally EMBL/ DDBJ and Genbank.  The databases differ in their approach to  sequence classification. The RDP database uses the RDP classifier to assign taxonomy to 16S rRNA sequences [@wang2007naive]. SILVA and Greengenes inherit a sequence's taxonomic assignment from the source database (such as NCBI or EBI). SILVA provides a non-redundant database version in which no taxonomic classification contains sequences with  greater than 99% pairwise identity [@quast2012silva]. Although each database performs sequence quality checks, only Greengenes actively checks for chimeric sequence, which can negatively affect 16S taxonomic assignment [@desantis2006greengenes]. 


The National Centre for Biotechnology Information (NCBI) provides multiple databases, including the Sequence Read Archive (SRA) [@kodama2011sequence] for raw  high-throughput sequencing data, and the Genome database as an umbrella for draft and complete genomes.  Not all genome sequences in the NCBI Genome database have publicly available raw data  in the SRA, and only  10% of genomes in the database are closed or complete [@waters2018riboseed]. A survey of the SILVA 132 database shows that 9.5% of sequences come from draft genome assemblies; the vast majority (87%) are obtained as amplicon sequences (usually Sanger sequenced), and the remaining 2% come from complete genomes. A common weakness of draft assemblies from short-read sequencing is incorrect assembly of repeated rDNA regions of a genome, which may be collapsed/merged into a single rDNA. The resulting 16S rRNA  could in turn be incorporated into SILVA, GreenGenes, or RDP. This compromises the quality of 16S rRNA databases; such sequences should be treated with caution. Genome assemblies from short reads are prone to errors in rDNA regions, as the length of the region exceeds read lengths. PCR spanning the rDNA region, followed by Sanger sequencing, or the use of long-read technologies such as PacBio or Nanopore sequencing, can resolve these multiple copies but, as the majority of the data generated over the last two decades comes from short read sequences, fixing these collapsed regions remains a valuable goal [@land2015insights;@wagner_evaluation_2016].  


The correct re-assembly of multiple rDNA regions of draft genomes can be achieved using riboSeed, which uses a reference genome to help assemble the rDNA regions of a draft genome [@waters2018riboseed]. riboSeed exploits the observation that the flanking regions of the rDNA region are highly conserved within a taxon yet variable between rDNA copies in the same genome, by using targeted subassembly to correctly place each re-assembled copy of multi-locus rDNA repeats.  The knock-on effect of assembling multi-copy rDNA operons is the gain of highly-accurate 16S rRNA sequences, which can be incorporated into 16S rRNA databases.


Here, we present the results of using such an approach to augment existing 16S rRNA databases with newly-assembled sequences from the SRA data corresponding to pre-existing draft genomes. The additional sequences provide greater coverage of ASVs in publicly-available datasets, aiding efforts to understand microbial communities.



# Methods

## The Focus16 Pipeline


We developed `r toolname`: a pipeline to augment existing 16S rRNA databases by mining the SRA database for candidate whole-genome sequencing studies for re-assembly. Candidate SRAs are identified, downloaded, subjected to automated quality control, re-assembled with riboSeed to resolve the rDNA operons, given a taxonomic assignment with Kraken, and formatted for addition to existing databases. Kraken uses the genome assembly as the basis for taxonomic assignment, mitigating the risk of misclassification when using a 16S rRNA classifier.


The  pipeline is shown as a flowchart (Figure \@ref(fig:flowchart-fig)). Details of third-party tools used in `r toolname` can be found in the Supplementary Methods section "Third-party software".

Given a genus or "Genus species" binomial, the pipeline progresses as follows:

1.  Candidate reference genomes are identified and downloaded from Refseq by matching the provided organism name (this may also be restricted to genomes defined by the user). 
2. Barrnap [@seemann_barrnap_2020]  is used to screen these complete genomes by estimating the 16S rRNA count.  Reference genomes with a single 16S rRNA are discarded. This catches two cases: 

    a. An organism may only have a single 16S rRNA. In this case riboSeed assembly will not improve on existing draft genomes.
    b.  A draft genome may have been incorrectly attributed the classification of "complete".  Errant reference genomes are therefore discarded, and the remaining references are available for use in the pipeline. 

3. A Mash [@ondov_mash_2016a] sketch is generated from the references passing the filtering in step 2.
4. sraFind ^[https://github.com/nickp60/sraFind] is used to identify all whole-genome sequencing SRA accessions for the organism of interest; these are downloaded with fastq-dump or fasterq-dump [@sra-tools_2019]. Steps 5a-5e proceed for each SRA.

    a. **Identify closest reference genome**. For a given SRA, the most compatible reference genome is determined via plentyofbugs [@waters_plentyofbugs_2019], which performs an initial assembly with the fast and highly-accurate assembler SKESA [@souvorov_skesa_2018] using a subsample of 1M reads. Mash is used to identify the closest match between the preliminary assembly and all the reference genomes from step 1. If no close match above a user-defined threshold (defaulting to a Mash distance of 0.1, roughly corresponding to a within-genus match [@ondov_mash_2016a]), the SRA is skipped; otherwise, the closest match is  later used for a reference with riboSeed. 
    b. **Classify Assembly**. Kraken2 [@woodKrakenUltrafastMetagenomic2014] is used to assign taxonomic classification to the preliminary SKESA assembly. The highest-ranked binomial name is recorded, and the full Kraken report is stored in the output folder for inspection.  
    c. **Pre-Assembly Quality Control**. Reads are run through several quality control steps. The average length of the reads is checked; an SRA that contains reads of very low length (i.e. less than 65bp) will be rejected, as very short reads cannot be used effectively by riboSeed to differentiate rDNA flanking regions. Reads are then quality trimmed with Sickle [@joshi_sickle_2011] using default parameters. fastp [@chen_fastp_2018] is used to identify and remove any remaining adapter sequences. For paired-end runs, unpaired reads are rejected.  
    d. **Downsample**. Read coverage is assessed using either a user-provided estimate of genome length or the length of the reference genome. If read coverage exceeds a user-specified threshold (defaults to 50x coverage, as further coverage can artificially support sequencing errors; see SPAdes paper [@bankevichSPAdesNewGenome2012]), trimmed reads are down-sampled to reach the desired coverage with seqtk [@li_seqtk_2020].
    e.  **De fere novo Assembly**. The SRA reads are then assembled using riboSeed, using the reference genome determined in (2a) as a template genome. Subassemblies are performed with SPAdes [@bankevichSPAdesNewGenome2012]; the default parameters of 3 rounds of seeding and 1kbp flanking regions are used.
    
5.  16S rDNA sequences are extracted and formatted.

    a.  Barrnap is run on either the subassemblies ("fast" mode) or from the final assembly ("full" mode).
    b.  Taxonomy assigned by Kraken2 in step 4b is used to label extracted sequnces.
    c.  Sequences are written to a fasta file that matches the format used by the SILVA database. 

In our analysis, we filtered to retain full-length sequences by removing any 16S rRNA under 1358bp (under the 1st quartile of the sequence lengths in SILVA). Additionally, we removed any 16S rRNA for which Kraken2's taxonomic report showed inconclusive domain-level taxonomic  assignment: assemblies were excluded if one domain was not assigned to over 70% of contigs. This was done to remove datasets that may be suffering from contamination (see Supplementary Methods section "Identifying poor taxonomic assignments", Figures S4 and S5).


```{r flowchart-fig, fig.cap="Flowchart of the pipeline that resolves multi-copy 16S loci from sequenced genomes with reads in SRA (as implemented in Focus16). Candidate reference genomes are downloaded from RefSeq. Reads for each SRA are downloaded and Kraken2 is ues to assign taxonomy. Corresponding reference genomes and SRA read sets are identified (using SKESA and Mash), and a new assembly constructed from these using riboSeed to resoolve 16S rDNAs. The assembled 16S rRNA regions are then taken forward for phylogenetic reconstruction, or to supplement existing reference databases. Number refer to stages outlined in the text; gray lines signify to taxonomic information, and black lines signify to sequence information.", out.width = '80%'}
knitr::include_graphics("figures/1-flowchart.png")
```

### Implementation

The `r toolname` pipeline can be installed from PyPI or from the source hosted on GitHub at <https://github.com/FEMLab/focus16>; all of the dependencies can be easily managed with Conda for reproducibility.  It was designed to efficiently handle the downloading and re-assembly of large amounts of short-read data. Users can use SRA-tools's prefetch command for faster downloads of SRA data; the re-assembly status of each SRA is recorded in an SQLite database. For those with access to a computing cluster running Open Grid Scheduler, the time-consuming  assembly steps can be distributed as array jobs as needed.  

The first time the pipeline is used, an automated setup procedure is run to download the required databases for Kraken2 and sraFind. 

Throughout, diagnostic information is recorded; if an aspect of the pipeline fails, rerunning the same command will reuse available intermediate results wherever appropriate.  

## Selecting suitable test datasets and identifying genera

Three mock communities described in the DADA2 manuscript  [@callahanDADA2HighresolutionSample2016] were selected to assess the utility of `r toolname`.  These communities, named "Extremes", "HMP", and "Balanced", and comprising 27, 21, and 57 members respectively [@schirmer_insight_2015;@kozich_development_2013;@callahanDADA2HighresolutionSample2016], were sequenced on an Illumina MiSeq yielding over 500,000 250bp paired-end reads. 

To provide an assessment of real-world usage of `r toolname`, we used the data generated in the EndoBiota study [@ataEndobiotaStudyComparison2019]^[PRJEB26800]: a survey of microbiomes across three body sites between women with and without endometriosis. These datasets are summarized in Table \@ref(tab:dataset-tab).

```{r dataset-tab, size="tiny", eval=TRUE}
knitr::kable(
  col.names = c("Name",	"N Strains",	"Reference",	"Description"),
  booktabs=TRUE,
  caption="Description of the four datasets considered in this study.  The number of strains was readily available for the mock communities; for the Endobiota study, this was determined by a preliminary analysis using DADA2 and SILVA. Asterisk (*) indicates that this is a calculated number, not a value known beforehand.",
  read.table(text="
Balanced	57	Schirmer et al. 2015	Mock community of bacteria and archaea; strains in equal proportions
HMP	21	Kozich et al. 2013	Mock community with common members of the human microbiome
Extremes	27	Callahan et al. 2016	Mock community of bacteria with greatly varying proportions
Endobiota	292*	Ata et al. 2019	Study comparing microbiomes of women with and without endometriosis
", sep="\t")
) %>% kable_styling(latex_options = "striped")
```



Unlike the mock communities, the number of genera present in the Endobiota samples is not known *a priori*. We estimated the abundances of community members by processing the samples through DADA2 in a similar manner to how the Balanced, Extremes, and HMP datasets had been analyzed. The analysis script `DADA2_analysis.Rmd` can be found in the supplementary materials repo; processed data was deposited  along with the rest of results generated in this work at  <https://zenodo.org/record/1172783>.  In short, DADA2 was used to build error profiles for each of the samples in the study. Reads were then trimmed 30bp on the 5' end and 40bp on the 3' end, quality trimmed after two low-Q bases, and any residual phi-X sequence was removed. Merged amplicons were filtered to retain those between 360 and 450 bases.  4.2% of sequences were determined to be chimeric and removed.  Taxonomy was assigned to the remaining 16S sequences using  DADA2's `assignTaxonomy` command with the SILVA non-redundant version 132, and species-level taxonomy was assigned using DADA2's `addSpecies` command as described in their manual.


In total, 333 unique genera were identified across the four datasets; these were cross-referenced with sraFind and RefSeq, filtering to retain those with both short-read SRAs available and at least one reference genome for that genus (Supplementary Figure S2, and Supplementary Table S2).
 
 The resulting list of 85 genera was used as the input for the pipeline.


## Assembly mode parameter choice
As candidate 16S rRNA sequences for a given organism could be identified from both riboSeed's subassemblies and the final *de fere novo* assembly, we sought to determine which of the two could be used. To do this, we determined the error (SNP/indel) rates by identifying which SRAs in our study had completed genomes, and performed an analysis comparing complete genomes to the sequences recovered from either *de novo* re-assembly of the complete genome, Focus16's "fast" mode (sequences from riboSeed's subassemblies), and "full" mode (sequences from riboSeed's final *de fere novo* assembly). The *de novo* assemblies were accurate but failed to recovery many individual 16S rDNAs, as is expected due to the repeated nature of these regions. riboSeed's subassemblies have both low error rates and successfully reconstruct the most rDNAs (Figure \@ref(fig:rdnamodes)), and as such are the ones we report below and recommend for augmenting a database.  
 

```{r rdnamodes, fig.cap="Comparing assembly modes for accuracy. SRAs in our dataset that underwent genome completion were used to identify the most accurate method of 16S rRNA sequence assembly. De novo assembly resulted in highly accurate 16S rRNA sequences, but was only able to recover 66 sequences. '--fast' mode proved to be the best tradeoff between accuracy and efficiency.", out.width = '80%'}
knitr::include_graphics("figures/S5-set-accuracy.png")
```

## Running `r toolname`
`r toolname` was run in a conda environment (see supplementary repository file "`r toolname`_env.txt"). Due to computational limitations (namely storage available on the high-performance computing cluster as well as the RAM required for genome assembly), we limited the scope of the analysis to a maximum of 50 randomly selected SRAs for each of the 85 genera. These are listed in the supplementary file `sras.tab`  The number of candidate reference genomes to be considered for each genus was capped at 200; the median number of genomes per genus was 9 (see Figure S3). The only genus with more than 200 reference genomes available was *Bordetella*. A minimum Mash distance was set to 0.1 [@ondov_mash_2016a] between a preliminary assembly and a reference genome, as this was shown to be the maximum distance between the reference and sequenced isolate that riboSeed performs well with (see @waters2018riboseed Figure 5). Run scripts are available in supplementary data; the reference genomes considered can be found  in Supplementary file `reference_genomes.tab`.


## Benchmarking Re-assembled 16S rRNA against Complete Genome 16S rRNA sequences {#benchmarking}

sraFind was used to identify which SRA accessions corresponded to complete NCBI genomes. For the genera considered in this study, these were matched with SILVA sequences sourced from complete genomes (see supplementary data "Provenance of strains"). Pairwise alignments were generated between the riboSeed 16S rRNA sequences and the SILVA sequences using the Biostrings package [@pages_biostrings_2020] in "overlap" mode (a global alignment with free ends) with a simple scoring matrix  (matches=1, mismatches=0); the highest-scoring alignment for each given reference 16S rRNA was used to identify misassemblies relative to the complete genome's 16S rRNA sequence. Alignments shorter than 1400bp were rejected.

## Benchmarking Re-assembled 16S rRNA against Draft 16S rRNA sequences {#benchmarking-draft}

Similar to the comparison to complete genomes above, we identified the SILVA sequences sourced from draft (see supplementary data "Provenance of strains"). Pairwise alignments were generated between the riboSeed 16S rRNA sequences and the SILVA sequences using the Biostrings package [@pages_biostrings_2020] in "overlap" mode (a global alignment with free ends) with a simple scoring matrix  (matches=1, mismatches=0); the highest-scoring alignment for each given reference 16S rRNA was used to identify missassemblies relative to the complete genome's 16S rRNA sequence. Alignments shorter than 1400bp were rejected.


## Assessing Taxonomic Assignment
The DADA2 pipeline was used to process each of the four datasets  in Table \@ref(tab:dataset-tab).  The resulting sequence tables were combined, and we assigned taxonomy with the naive Bayes classifier implemented in DADA2. This classified at the genus level, and DADA2's "assignSpecies" command was used to assign species-level taxonomy; we enabled the "allowMultiple" parameter to view ambiguities in the assignment. This analysis compares assignment with both SILVA 132 and assignment with SILVA 132 augmented with sequences generated by `r toolname`. All scripts can be found in supplementary materials. 

 
# Results

##  Benchmarking Re-assembly Accuracy 
riboSeed has been shown to generate high-quality reconstructions of each rDNA region when benchmarked against hybrid assemblies [@waters2018riboseed]. Using sraFind, we identified which sequences in SILVA originated from closed, complete genomes; those with short-read SRAs were used to benchmark the accuracy of the 16S rRNA sequences re-assembled with the `r toolname` pipeline (as described in Methods section \@ref(benchmarking)) against the sequence in SILVA. In our dataset, 61  of these SRA/complete genome pairs were present. 

```{r accuracy, echo=FALSE, fig.cap=paste0("Assembly errors per kilobase calculated between each ", toolname, " sequence and the reference genome as counts (A) and cumulatively (B). In the genera considered, 285 sequences from 61 complete genomes were present in SILVA; riboSeed recovered 271 of these. 146 of these 16S alleles were identical between riboSeed and complete genome (grey bars)."), out.width = '80%'}
knitr::include_graphics("figures/4-fast-cg-accuracy.png")
```

Comparing the re-assembled 16S rRNA sequences to the 16S rRNA sequences from complete genomes shows a low error rate, with 53% of sequences being perfect reconstructions and 95% of sequences having fewer than X errors  (Figure \@ref(fig:accuracy)). This confirms that `r toolname`'s best-case accuracy yields perfect reconstructions of the rDNA region; those cases for which reconstruction was imperfect rarely have more than 10 errors (an error rate rarely exceeding 0.7%), and 99% of sequences had fewer than 10 errors per Kbp. This suggesting that sequences could be used to augment existing databases.  The error rates for amplicon data in SILVA are difficult to determine; in a perfect sequencing run Sanger sequencing has vanishingly small error rates; where multiple sequences are inadvertently sequenced at the same time (i.e. multiple copies from a single organism), the trace will reflect the differences as short or imperfect peaks.  As the trace/quality data for amplicon sequences are not typically available, it is impossible to determine the accuracy of such sequences.

##  Comparing re-assembled 16S rRNA to draft 16S rRNA
```{r draftexamples, out.width="100%", fig.cap="Representative SNP alignments comparing 16S rDNA seqeunces from original draft assemblies to the re-assembled seqeunces. Alignments were generated with DECIPER and all columns matching the consensus were removed; original alignment length and column numbers for each SNP are shown under the sequence names.  Names starting with an SRA accession such as ERR and SRR are the re-assembled sequences.  The following types of relationships occur:  all alleles recovered in original and re-assembled (AB), sole original sequence misses a single (C) or multiple SNPs (D), dissagreement between sole original and re-assembled alleles (E), original sequences appears to be amalgamation of alleles (FG), a deletion is present in original allele (H), re-assembly fails to reconstruct an allele (I)."}
knitr::include_graphics("figures/draftaln.png")

```

As repeated rDNA operons are difficult to resolve with short read sequences, draft genome assemblies can (and often do) contain a single assembled rDNA region with elevated read coverage. This can be problematic for genus or 16S rRNA classification as– the 16S rRNA recovered may not just correspond to one of several 16S rRNA copies, but it can be a consensus "summary"/"collapsed" 16S rRNA resulting from imperfect assembly of the repeated region. We provide a few examples of such alignments in Figure  \@ref(fig:draftexamples) (see all alignments in supplementary repo folder  `figures/draft_alignments/`) .



In such cases, without the capacity to verify the regions with Sanger or long-read sequencing, determining which sequences are the missassemblies and which should be regarded as true is an impossible question to answer.


##  Augmenting SILVA with results from `r toolname`
### Recovering Sequences from Re-assembly
<!-- (Figure \@ref(fig:sras)) -->
`r toolname` was used to build an extended database for the three mock datasets described in the DADA2 paper and a real-world dataset from the Endobiota study. From the 85 genera considered, `r toolname` processed 2387 SRAs, and recovered 16S rRNA sequences from 1392 SRAs. The average execution time for a given SRA was approximately 23 minutes. Several factors can contribute to failing to recover 16S rRNA sequences from a given SRA; among these are a too-distant reference genome, low rDNA flanking diversity, low read length, or high read error rates. In total, we recovered 5854 16S rRNA sequences, of which 3008 were unique. 


### Recovery of unique sequences 
Ideally, `r toolname` would be applied to every eligible SRA currently available, and periodically rerun as more high-quality reference genomes are generated with long-read technologies; in this pilot study, we assessed the increase in unique sequences gained by augmenting SILVA with the 85 genera considered in this study. For thoroughly-sequenced genera such as *Escherichia*, *Pseudomonas*, or *Bacillus*, the increases in unique sequences  are small. However, other taxa showed marked increases in the genus-level diversity (Figure \@ref(fig:reassembly)). 


 ```{r reassembly, echo=FALSE, fig.cap="Green circles indicate SRAs that yielded 16S rRNA sequences, while blue diamonds indicate SRAs failing to yield re-assembled 16S rRNA sequences. Pink triangles show SRAs that were rejected due to limitations in the diversity of available reference genomes, and inverted green triangles show SRAs rejected due to read length, insufficient coverage, poor read quality, etc. A few errors occurred, usually when the SRAs metadata conflicted with the associated sequencing data and caused download errors or errors from reads with incorrect pairing.  In these cases, the data is discarded.", out.width = '80%'}
knitr::include_graphics("figures/2-genera.png")
```

## Assessing Taxonomic Assignment 
DADA2 was used to identify ASVs from the 4 datasets, resulting in a total of 4098 ASVs (109 sequences from the "HMP" dataset, 26 from the "Extreme" dataset, 94 from the "balanced" dataset, and the rest from the EndoBiota study). We then compared the taxonomic results of classification using either the SILVA 132 or SILVA 132 augmented with the `r toolname`. 

Of the 4098 sequences, `r toolname` changed the taxonomic assignment of 20 strains; see Table 2, or Supplementary file STABLE_different_assignment.tab for the actual sequences. Changes could happen in two ways: a previously unclassified ASV gained classification, or previously-assigned ASV gained more species- (or genus-) level details. In our dataset, three unclassified strains gained annotations (numbers 5, 10, and 11) . The remaining 17 had more detail added to the genus or species level; usually, this meant that with SILVA alone, a single species classification was given, but with the augmented database, it was revealed that the ASV could belong to more than one other organisms at either the species (numbers 4,6,8,9,12,13,14 ) or genus (1,2,7, 15-20) level.

It is impossible to accurately estimate the expected gains of a wider application of this approach. If this study is perfectly representative, users could expect an information gain of about 0.5%; two factors must be considered before extrapolating that value beyond this study.  First, we were limited in the number of SRAs per genus that could be processed; Supplementary Figure 1 shows the number of SRAs per genus. Second, and perhaps more importantly, the majority of genera in this study are associated with the human microbiomes, an area which has already seen an extensive amount of focus in terms of amplicon sequencing, whole-genome sequencing, and genome completion efforts. Other environments have not had this benefit, and perhaps have more room for improvement.

```{r diversity, echo=FALSE, fig.cap=paste0("Percentage increase in unique sequences (and number of added unique sequences), by genus. The addition of the sequences recovered by ", toolname, " increases the number of unique sequences for the given taxa; the greatest increases are found in \"under-sequenced\" taxa and/or taxa with less well-conserved rRNA sequences."), out.width = '80%'}
knitr::include_graphics("figures/3-perc_unique_genus.png")

```

```{r hits-tab, size="tiny"}
knitr::kable(
  booktabs=TRUE,
  caption="Of the 4098 ASVs aggregated across the four datasets considered, the augmented database allowed improved taxonomic assignment of 20 sequences. Cells reading 'NA NA' indicate that no genus or species could be assigned. Cells reading NA speciesA/speciesB indicate that the lowest common ancestor for a sequence could not be determined at the genus level; in such cases DADA2 gives the designation NA rather than listing possible genera as is done at species level.", col.names =  c("", "SILVA", paste("SILVA + ", toolname)),
  read.csv("../STABLE_different_assignment.tab", sep="\t") %>% 
    mutate(set=row_number()) %>% 
    select(set, silva, focuDB) %>% 
    transform(
      silva =gsub("\\/", "\\/ ", silva),
      focuDB =gsub("\\/", "\\/ ", focuDB)) 
) %>% column_spec(1, width = "2em")   %>% column_spec(2, width = "20em")  %>% column_spec(3, width = "20em")  %>% kable_styling(latex_options = "striped")

 # kable_styling(full_width = F) %>%
 #  column_spec(1, bold = T, border_right = T) %>%
 #  column_spec(2, width = "30em", background = "yellow")

```


# Discussion
`r toolname` orchestrates the re-assembly of whole-genome sequencing datasets in SRA to recover 16S rRNA sequences that may be missing from the existing reference databases. Using riboSeed, `r toolname` re-assembles draft genomes that currently contribute a single (often collapsed) 16S rRNA to resolve distinct instances of the 16S rRNA operon. We show that this increases the sequence diversity (number of unique sequences) of the 16S rRNA databases, and that the increased diversity results in measurable improvements to taxonomic assignment.

`r toolname` improved fine-scale taxonomic assignment in two ways: by assigning previously unclassified sequences, and by revealing "overeager" species assignment when a 16S rDNA sequence could have come from two or more species. While at face-value this decreases the precision of taxonomic assignment, it reveals cases where species-level assignment was inappropriate.   Based on the information gain across the 85 genera considered here, we believe a wide-scale application of `r toolname` could benefit the community.

<!-- ((98329*(23/60)) /24)/150 -->
However, there are three main limitations facing the large-scale application of Focus16: the first is the bandwidth, computational power, memory, and storage required to re-assemble the 98,329 SRAs (as of October 2019) that were used to generate draft genomes. Given sufficient storage and unfettered access to a medium/large computing cluster such as one supporting a university or research institute (say 150 compute nodes), the task could be completed within two weeks^[This estimation is based on an rough average processing time of 23 minutes per run, but this is highly dependant on download speed, read/write speed, genome size, and sequencing depth.]; however with modest hardware (8 cores, 20GB RAM), this would take about 4 years in “--fast” mode. These estimates are ignoring the 112,695 draft genomes for which no reads were ever released, which leads to the second limitation: data availability. Releasing draft genomes without the reads used to generate them hampers efforts such as this one to expand beyond the purpose of the original study.

The third of this approach is the availability of high-quality closed genomes to use as references. With the increased adoption of long read technologies, we envisage that this limitation will decrease with time; re-running the pipeline as new, complete reference genomes are generated will allow for ongoing improvements to the databases. Eventually, a point will come when `r toolname` will no longer be needed as all candidate SRAs have been sufficiently utilized.

Further limitations exist within `r toolname` and within riboSeed. The success of riboSeed's *de fere novo* assembly hinges on the similarity of the reference to the sequenced isolate, the differentiating power of the rDNA flanking regions, read length, and other factors. This is one reason why not all SRAs yielded perfect rDNAs. Additionally, riboSeed does not currently support mate-paired libraries; these are much less widely used than the typical single-end or paired-end libraries used in short-read sequencing, but represents a limitation none the less.

Despite these limitations, we have shown that `r toolname` can contribute towards better molecular ecology analysis;  augmenting SILVA with the sequences re-assembled from the 85 genera considered here led to a small increase in the number of unique sequences in the database.  Using the augmented database for taxonomic assignment revealed some limitations of low-level taxonomic assignment, and led to the classification of a few additional ASVs.   We invite the community to consider augmenting existing databases (such as NCBI's 16S RefSeq Microbial database, SILVA, RDP, and GreenGenes) with the approach outlined here.



# Competing interests
The authors declare that they have no competing interests.

# Funding
This work was financially supported by Science Foundation Ireland (Awards 14/IA/2371 and 16/RC/3889) and through through a joint studentship between The James Hutton Institute and the National University of Ireland, Galway.

# Acknowledgements
Many thanks Christopher Quince for the helpful conversations on the topic.

# Data Accessibility
The code for Focus16 can be found at <https://github.com/FEMLab/focus16>; the code for all the analyses presented in this work can be found at <https://github.com/FEMLab/focus16_manuscript>. All data used is archived at Zenodo accession 10.5281/zenodo.3956433.

# Author Contributions
Author contributions according to the CRediT taxonomy [@allen_how_2019] are listed alphabetically as follows:
Conceptualization: LP, NW ; Methodology: FA, BN, LP, NW; Software and Data Curation: BN, NW; Validation: BN, LP, NW; Formal analysis: BN, LP, NW; Investigation: BN, NW; Resources; FA, VOF, LP; Writing - Original Draft: BN, NW; Writing - Review & Editing: FA, FB, VOF, AH,  BN, LP, NW; Visualization: BN, LP, NW; Supervision: FA, FB, VOF, AH, LP, NW; Project administration	FA, VOF, NW; Funding acquisition: FA, FB, VOF, LP, NW.


# References
