---
title: 'Mining underutilized whole-genome sequencing projects to improve 16S rRNA databases'
author:
  - name: Ben Nolan
    affiliation: [fem, mel]
  - name: Florence Abram 
    affiliation: fem
  - name: Fiona Brennan
    affiliation: sem
  - name: Ashleigh Holmes
    affiliation: cms
  - name: Vincent O’Flaherty 
    affiliation: mel
  - name: Leighton Pritchard
    corresponding: true
    affiliation: [ics, strath]
  - name: and Nicholas R. Waters
    email: leighton.pritchard@strath.ac.uk or watersn@mskcc.org
    corresponding: watersn@mskcc.org
    affiliation: [fem,  sem, ics]

address:
  - code: fem
    address: Functional Environmental Microbiology, Microbiology, School of Natural Sciences, National University of Ireland, Galway, Ireland
  - code: mel
    address: Microbial Ecology Laboratory, Microbiology, School of Natural Sciences, National University of Ireland, Galway, Ireland
  - code: sem
    address: Soil and Environmental Microbiology, Environmental Research Centre, Teagasc, Johnstown Castle, Wexford, Ireland
  - code: cms
    address: Cell and Molecular Sciences, James Hutton Institute, Invergowrie, Dundee DD2 5DA, Scotland
  - code: ics
    address: Information and Computational Sciences, James Hutton Institute, Invergowrie, Dundee DD2 5DA, Scotland
  - code: strath
    address: Strathclyde Institute of Pharmacy and Biomedical Sciences, University of Strathclyde, Glasgow, G1 1XQ, Scotland

date: "`r format(Sys.time(), '%d %B, %Y')`"
toc: false
fontsize: 10pt
output:
  bookdown::pdf_document2:
     fig_caption: true
     keep_tex: true
     template: "template.tex"
     #latex_engine: xelatex
editor_options:
  chunk_output_type: console
bibliography: focusdb_references.bib
abstract: Current approaches to interpreting 16S rDNA amplicon data are hampered by several factors. Among these are database inaccuracy or incompleteness, sequencing error, and biased DNA/RNA extraction. Existing 16S rRNA databases source the majority of sequences from deposited amplicon sequences, draft genomes, and complete genomes. Most of the draft genomes available are assembled from short reads. However, repeated ribosomal regions are notoriously difficult to assemble well from short reads, and as a consequence the short-read-assembled 16S rDNA region may be an amalgamation of different loci within the genome. This complicates high-resolution community analysis, as a draft genome's 16S rDNA sequence may be a chimera of multiple loci; in such cases, the draft-derived sequences in a database may not represent a 16S rRNA sequence as it occurs in nature. We present Focus16, a pipeline for improving 16S rRNA databases by mining NCBI's Sequence Read Archive for whole-genome sequencing runs that could be reassembled to yield additional 16S rRNA sequences. Using riboSeed (a genome assembly tool for correcting rDNA misassembly), Focus16 provides a way to augment 16S rRNA databases with high-quality re-assembled sequences. In this study, we augmented the widely-used SILVA 16S rRNA database with the novel sequences disclosed by Focus16 and re-processed amplicon sequences from several benchmarking datasets with DADA2. Using this augmented SILVA database increased the number of amplicon sequence variants (ASVs)that could be assigned taxonomic annotations. Further, fine-scale classification was improved by revealing ambiguities. We observed, for example, that some amplicon sequence variants (ASVs) originally assigned to a specific genus were indicated by Focus16-correction to be ambiguous, as the ASV was represented in two or more genera.  Thus, we demonstrate that improvements can be made to taxonomic classification by incorporating these carefully re-assembled 16S rRNA sequences, and we invite the community to expand our work to augment existing 16S rRNA reference databases such as SILVA, GreenGenes, and RDP.
---


<!--     #affiliation: Information and Computational Sciences, James Hutton Institute -->
<!--     #address: Invergowrie, Dundee DD2 5DA, Scotland -->
<!-- #     - %\author{\protect\parbox{\textwidth}{\protect\centering\small Ben Nolan,$^{1}$ Florence Abram,$^{1}$ Fiona Brennan,$^{1,3}$ Ashleigh Holmes,$^{4}$  Vincent O’Flaherty$^{2}$,\newline Leighton Pritchard$^{5,6\ast}$ and Nicholas R Waters$^{1,5}$ \newline \newline \small{\textit {$^{1}$Functional Environmental Microbiology, School of Natural Sciences, National University of Ireland, Galway, Ireland}}  \small{\textit $^{2}$Microbial Ecology Laboratory, Microbiology, School of Natural Sciences, National University of Ireland, Galway, Ireland}} \newline \small{\textit {$^{3}$Soil and Environmental Microbiology, Environmental Research Centre, Teagasc, Johnstown Castle, Wexford, Ireland}}\newline \small{\textit {$^{4}$Cell and Molecular Sciences, James Hutton Institute, Invergowrie, Dundee DD2 5DA, Scotland}}\newline  \small{\textit {$^{5}$Information and Computational Sciences, James Hutton Institute, Invergowrie, Dundee DD2 5DA, Scotland}} \newline \small{\textit {$^{6}$Strathclyde Institute of Pharmacy and Biomedical Sciences, University of Strathclyde, Glasgow, G1 1XQ, Scotland}}\newline \small{\textit {$^{2,5}$Information and Computational Sciences, James Hutton Institute, Invergowrie, Dundee DD2 5DA, Scotland}} \newline \footnotesize{$^\ast$To whom correspondence should be addressed$:$ nick.waters@bioskryb.com or leighton.pritchard@strath.ac.uk \newline \newline}} -->

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE)
knitr::opts_chunk$set(cache=FALSE)
knitr::opts_chunk$set(fig.pos = 'H')
library(tidyverse)
library(diagram)

library(kableExtra)
# draft 4 but I think we have a name
toolname <- "Focus16"
options(tinytex.verbose = TRUE) 

```


# Introduction
The use of genetic markers for microbial classification has seen explosive growth over the past decade   [@liu_popular_2012;@boers_understanding_2019]. The 16S rRNA gene is the standard utilised in the assessment of prokaryotic community composition by amplicon sequencing [@fukuda_molecular_2016]. 16S rRNA has been used for community analysis in diverse environments such as the gut microbiota of cattle and pigs  [@avila-jaime_prediction_2018], soil [@santamaria_soil_2018], marine environments   [@dang_bacterial_2000], and the human gut [@jovel_characterization_2016]. The success of this method hinges on the presence of the 16S rRNA gene in all bacteria and archaea and its relatively slow rate of base substitution; thus the rDNA regions can be targeted with primers but the amplicon sequences exhibit enough diversity that organisms can be differentiated at the genus or species level [@woese_natural_1990;@woo_then_2008]. 

There are both technical and biological limitations of assessing microbial communities via 16S rRNA sequencing. Technical limitations include the compositional nature of sequencing data, extraction biases, and primer choice biases [@acinas_pcrinduced_2005,@giangacomo_comparing_2021,@gloor_microbiome_2017.  Biologically, 16S rRNA sequences are not a perfect proxy for taxonomy [@parks_standardized_2018], and 16S rRNA copy number  is difficult to account for.

Microbial genomes have a range of 16S rRNA gene copy numbers (GCNs) ranging from (for instance) Mycobacteria with a single copy to  *Photobacterium damselae* Phdp Wu-121 with 21 copies [@vetrovsky_variability_2013;@stoddard_rrndb_2014;@lee_rrndb_2009,@acinas_divergence_2004]. Organisms often harbor multiple 16S rRNA alleles [@sun_intragenomic_2013]; this can negatively impact 16S rRNA classification in two ways. First, in taxa with low or no 16S rRNA sequence variability, relative abundance estimates may be  overestimated in taxa with higher GCN and underestimated those with low GCN. For example, in a community of two organisms where one has five rDNA copies and the other only has a single copy, evenly sequencing the community would show a one-to-five abundance ratio. While methods have been developed to attempt to correct this, it remains a problem due to the tools heavy reliance on well-curated references [@louca_correcting_2018]. Second, some organisms have sufficient sequence variability between copies that they may be assigned different taxonomic classifications; indeed, certain extremophiles have been reported to possess very high 16S rRNA copy heterogeneity, where pairwise sequence identity of some 16S rRNA sequences on the same replicon can be as low as 91%  [@sun_intragenomic_2013]; this is well below the 97% or 99% clustering thresholds commonly used for community analysis, and certainly below the zero-radius Operational Taxonomic Unit boundary [@edgar_updating_2018]. @schloss_amplicon_2021 demonstrates clearly that within-genome heterogeneity is widespread, and estimated a roughly 2:1 ratio of copies to distinct 16S alleles.

These intragenomic rDNA repeats complicate community analysis, but each instance of the 16S rRNA contains valuable information. An ideal community analysis framework would utilize a database incorporating this information to both correct for copy number variation between organisms in a community, and correctly assign 16S rRNA alleles to the appropriate organism. 


Amongst the most widely used 16S rRNA databases for bacteria and archaea are Greengenes [@desantis_greengenes_2006], SILVA [@quast_silva_2012], and the Ribosomal Database Project (RDP) [@cole_ribosomal_2003]. Each contains 16S rRNA sequences derived from multiple major international nucleotide sequence databases, principally EMBL/ DDBJ and Genbank.  The databases differ in their approach to  sequence classification. The RDP database uses the RDP classifier to assign taxonomy to 16S rRNA sequences [@wang_naive_2007]. SILVA and Greengenes inherit a sequence's taxonomic assignment from the source database (such as NCBI or EBI). SILVA provides a non-redundant database version in which no taxonomic classification contains sequences with  greater than 99% pairwise identity [@quast_silva_2012]. Although each database performs sequence quality checks, only Greengenes actively checks for chimeric sequence that can negatively affect 16S taxonomic assignment [@desantis_greengenes_2006]. 


The National Center for Biotechnology Information (NCBI) provides multiple databases, including the Sequence Read Archive (SRA) [@kodama_sequence_2012] for raw  high-throughput sequencing data, and the Genome database containing draft and complete genomes.  Not all genome sequences in the NCBI Genome database have publicly available raw data in the SRA, and the minority of those genomes in the database are closed or complete [@waters_riboseed_2018]. Tabulating the accession types of the SILVA 132 database shows that 9.5% of sequences come from draft genome assemblies; the vast majority (87%) are obtained as amplicon sequences (usually Sanger sequenced), and the remaining 2% come from complete genomes. While Sanger-sequenced amplicons are generally very accurate, a common weakness of draft assemblies from short-read sequencing is incorrect assembly of repeated rDNA regions of a genome, which may be collapsed or merged into a single rDNA. The resulting 16S rRNA sequence could in turn be incorporated into SILVA, GreenGenes, or RDP. This compromises the quality of 16S rRNA databases, and  such sequences should be treated with caution. Genome assemblies from short reads are prone to errors in rDNA regions, as the length of the repeated region exceeds read lengths. PCR spanning the rDNA region, followed by Sanger sequencing, or the use of long-read technologies such as PacBio or Nanopore sequencing, can resolve these multiple copies but, as the majority of the data generated over the last two decades comes from short read sequences, fixing collapsed regions remains a valuable goal [@land_insights_2015;@wagner_evaluation_2016].  

Ideally, 16S rRNA reference databases would contain each allele present in each organism. In pursuit of that aim, we explored re-assembly of draft genomes as a means to obtain high-quality 16S rRNA alleles. We used riboSeed for reassembly, which exploits the observation that the flanking regions of the rDNA region are highly conserved within a taxon yet variable between rDNA copies within a genome; it uses a  targeted subassembly approach to resolve each re-assembled copy of multi-locus rDNA repeat [@waters_riboseed_2018]s. The resulting 16S rRNA alleles can expand the breadth of existing reference databases.

Here, we present the results of using this approach. We augmented the SILVA 132 database with new 16S rRNA alleles obtained from reassembly of draft genomes, and  found the additional sequences allow for better Amplicon Sequence Variant (ASV) taxonomic annotation, aiding efforts to understand microbial communities.



# Methods

## The Focus16 Pipeline


We developed `r toolname`: a pipeline to augment existing 16S rRNA databases by mining the SRA database for candidate whole-genome sequencing studies that could be reassembled with improved 16S sequence assembly accuracy. Candidate SRAs are identified, downloaded, subjected to automated quality control, re-assembled with riboSeed to resolve the rDNA operons, given a taxonomic assignment with Kraken, and formatted for addition to existing databases. Kraken2 assigns taxonomy using exact matches to a lowest common ancestor for k-mers from the whole genome assembly, mitigating the risk of misclassification compared to using a 16S rRNA classifier alone.


The  pipeline is shown as a flowchart (Figure \@ref(fig:flowchart-fig)). Details of third-party tools used in `r toolname` can be found in the Supplementary Methods section "Third-party software".

Given a genus or "Genus species" binomial, the pipeline progresses as follows:

1.  Candidate complete reference genome are identified and downloaded from RefSeq [@pruitt_ncbi_2005] by matching the provided organism name. Users can optionally provide additional reference genomes.
2. Barrnap [@seemann_barrnap_2020]  is used to screen these complete genomes by estimating the 16S rRNA count.  Reference genomes with a single 16S rRNA are discarded. This catches two cases:

    a. The taxon represented by this organism  may only have a single 16S rRNA. In this case riboSeed assembly will not improve on existing draft genomes.
    b.  A draft genome may have been incorrectly attributed the classification of "complete", and present as having a single rRNA sequence when it in fact has more than one.  Any such errant reference genomes are therefore discarded, and the remaining references are available for use in the pipeline.

3. A Mash [@ondov_mash_2016] sketch is generated from the references passing the filtering in step 2.
4. sraFind (https://github.com/nickp60/sraFind) is used to identify all whole-genome sequencing SRA accessions for the taxon of interest; these are downloaded with fastq-dump or fasterq-dump [@_sratools_2019]. Steps 4a-4e are applied to each SRA.

    a. **Identify closest reference genome**. For each SRA, the most compatible reference genome is determined via plentyofbugs [@waters_plentyofbugs_2019], which performs an initial assembly with the fast and highly-accurate assembler SKESA [@souvorov_skesa_2018] using a subsample of 1M reads. Mash is used to identify the closest match between the preliminary assembly and all the reference genomes from step 1. If no close match below a user-defined threshold (defaulting to a Mash distance of 0.1, roughly corresponding to a within-genus match [@ondov_mash_2016]) is identified, the SRA is skipped; otherwise, the closest match is retained for use as a reference genome.
    b. **Classify Assembly**. Kraken2 [@wood_improved_2019] is used to assign taxonomic classification to the preliminary SKESA assembly. The highest-ranked binomial name is recorded, and the full Kraken2 report is stored in the output folder for inspection.
    c. **Pre-Assembly Quality Control**. Reads are run through several quality control steps. The average length of the reads is checked; an SRA that contains reads of very short length (i.e. an average less than 65bp) is rejected, as very short reads cannot be used effectively by riboSeed to differentiate rDNA flanking regions. Reads are then quality trimmed with Sickle [@joshi_sickle_2011] using default parameters. fastp [@chen_fastp_2018] is used to identify and remove any remaining adapter sequences. For paired-end runs, unpaired reads are rejected.
    d. **Downsample**. Read coverage is assessed using either a user-provided estimate of genome length or the length of the reference genome. If read coverage exceeds a user-specified threshold (defaults to 50x coverage, as further coverage can artificially support sequencing errors; see [@bankevich_spades_2012]), trimmed reads are down-sampled to reach the desired coverage with seqtk [@li_seqtk_2020].
    e.  **De fere novo Assembly**. The SRA reads (or downsampled reads from 4d) are then assembled using riboSeed, using the reference genome determined in (2a) as a template genome. Subassemblies are performed with SPAdes [@bankevich_spades_2012]; the default parameters of 3 rounds of seeding and 1kbp flanking regions are used.

5.  16S rDNA sequences are extracted and formatted.

    a.  Barrnap is run on either the subassemblies ("fast" mode) or the final assembly ("full" mode).
    b.  Taxonomy assigned by Kraken2 in step 4b is used to label extracted sequences.
    c.  Sequences are written to a FASTA file that matches the format used by the SILVA database.

In our analysis, we filtered to retain only full-length sequences by removing any 16S rRNA shorter than 1358bp (under the 1st quartile of the sequence lengths in SILVA). Additionally, we removed any 16S rRNA for which Kraken2's  report showed inconclusive domain-level taxonomic  assignment: assemblies were excluded if the same domain was not assigned to over 70% of contigs. This was done to remove potentially-contaminated datasets (see Supplementary Methods section "Identifying poor taxonomic assignments", Figures S4 and S5).


```{r flowchart-fig, fig.cap="Flowchart of the pipeline that resolves multi-copy 16S rRNA loci from sequenced genomes with corresponding reads in SRA (as implemented in Focus16). Candidate reference genomes are downloaded from RefSeq. Reads for each SRA are downloaded and Kraken2 is used to assign taxonomy. Corresponding reference genomes and SRA read sets are identified (using SKESA and Mash), and a new assembly constructed from these using riboSeed to resolve 16S rDNAs. The assembled 16S rRNA regions are then taken forward for phylogenetic reconstruction, or to supplement existing reference databases. Numbers refer to stages outlined in the text; gray lines signify taxonomic information, and black lines signify sequence information.", out.width = '80%'}
knitr::include_graphics("figures/1-flowchart.png")
```

### Implementation

The `r toolname` pipeline can be installed from PyPI or from the source hosted on GitHub at <https://github.com/FEMLab/focus16>; all of the dependencies can be easily managed with Conda for reproducibility.  The package was designed to efficiently handle the downloading and re-assembly of large amounts of short-read data. Users can use SRA-tools's prefetch command for faster downloads of SRA data; the re-assembly status of each SRA is recorded in an SQLite database. For those with access to a computing cluster running Open Grid Scheduler, the time-consuming  assembly steps can be distributed as array jobs as needed.

The first time the pipeline is used, an automated setup procedure is run to download the required databases for Kraken2 and sraFind.

Throughout, diagnostic information is recorded; if an aspect of the pipeline fails, rerunning the same command will reuse available intermediate results wherever appropriate.

## Selecting suitable test datasets and identifying genera

Three mock communities described in the DADA2 manuscript  [@callahan_dada2_2016] were selected to assess the utility of `r toolname`.  These communities, named "Extremes", "HMP", and "Balanced", and comprising 27, 21, and 57 members respectively [@schirmer_insight_2015;@kozich_development_2013;@callahan_dada2_2016], were sequenced on an Illumina MiSeq yielding over 500,000 250bp paired-end reads each. To provide an assessment of real-world usage of `r toolname`, we used the data generated in the EndoBiota study [@ata_endobiota_2019](PRJEB26800): a survey of microbiomes across three body sites of women with and without endometriosis. These datasets are summarized in Table \@ref(tab:dataset-tab).

```{r dataset-tab, size="tiny", eval=TRUE}
knitr::kable(
  col.names = c("Name",	"N Strains",	"Reference",	"Description"),
  booktabs=TRUE,
  caption="Description of the four datasets considered in this study.  The number of strains was readily available for the mock communities; for the Endobiota study, this was determined by a preliminary analysis using DADA2 and SILVA. Asterisk (*) indicates that this is a calculated number, not a value known beforehand.",
  read.table(text="
Balanced	57	Schirmer et al. 2015	Mock community of bacteria and archaea; strains in equal proportions
HMP	21	Kozich et al. 2013	Mock community with common members of the human microbiome
Extremes	27	Callahan et al. 2016	Mock community of bacteria with greatly varying proportions
Endobiota	292*	Ata et al. 2019	Study comparing microbiomes of women with and without endometriosis
", sep="\t")
) %>% kable_styling(latex_options = "striped")
```



Unlike the mock communities, the number of genera present in the Endobiota samples is not known *a priori*. We estimated the abundances of community members by processing the samples through DADA2 in a similar manner to how the Balanced, Extremes, and HMP datasets had been analyzed. The analysis script `DADA2_analysis.Rmd` can be found in the supplementary materials repository. Processed data was deposited along with the rest of results generated in this work at  <https://zenodo.org/record/1172783>.  In short, DADA2 was used to build error profiles for each of the samples in the study. Reads were then trimmed 30bp on the 5' end and 40bp on the 3' end, quality trimmed after two low-Q bases, and any residual phi-X sequence was removed. Merged amplicons were filtered to retain those of length between 360 and 450 bases.  4.2% of sequences were determined to be chimeric and removed.  Taxonomy was assigned to the remaining 16S rRNA sequences using  DADA2's `assignTaxonomy` command with the SILVA non-redundant version 132, and species-level taxonomy was assigned using DADA2's `addSpecies` command as described in their manual.


In total, 333 unique genera were identified across the four datasets; these were cross-referenced with sraFind and RefSeq, and filtered to retain only those with both short-read SRAs available and at least one reference genome for that genus (Supplementary Figure S2, and Supplementary Table S2). The resulting list of 85 genera was used as the input for the `r toolname` pipeline.


## Assembly mode parameter choice
As candidate 16S rRNA sequences for a given organism could be extracted from riboSeed's subassemblies or the final *de fere novo* assembly, and we sought to determine which was the best choice for generating sequence to extend the reference databases. Indeed, generating the subassembled sequences alone is a less computationally-intensive process than whole-genome assembly, but the final whole-genome assembly step acts as a further refinement of the subassemblies. To find which source is the best for augmenting 16S rRNA databases, we determined the error (SNP/indel) rates by comparing complete genomes to the sequences recovered from either *de novo* re-assembly of the complete genome, Focus16's "fast" mode (sequences from riboSeed's subassemblies only), and "full" mode (sequences from riboSeed's final *de fere novo* assembly). As most 16S rRNA sequencing studies target one or more hypervariable regions rather than the full-length gene, we used the DECIPHER package to extract amplicons from the full-length sequences for two popular targets: the V1-V3 region used by the Human Microbiome Project [@humanmicrobiomeprojectconsortium_structure_2012,humanmicrobiomeprojectconsortium_framework_2012], and the V3-V4 region recommended by Illumina [@klindworth_evaluation_2013, @_16s_].



```{r rdnamodes, fig.cap="Comparing assembly modes for accuracy and sequence recovery against complete genomes. SRAs in our dataset that were used to generate complete genomes were used to determine assembly mode accuracy. (A) De novo assembly resulted in highly accurate 16S rRNA sequences, but was only able to recover 66 of the 288 sequences in the complete genomes. riboSeed's '--full' and '--fast'  modes recovered 241 and 271 sequences, respectively. Results shown for full-length 16S as well as the V1-V3 and V3-V4 regions. Assembly name is indicated by color; those with over 10 errors pkb are labeled. The number of perfect sequences recovered is shown under the boxplot. (B) The cumulative distribution of errors across the three regions (fast mode). (C) Histogram of error rates by region  (fast mode)." , out.width = '80%'}
knitr::include_graphics("figures/new2-accuracy.png")
```

## Running `r toolname`
`r toolname` was run in a conda environment (see supplementary file "`r toolname`_env.txt"). Due to computational constraints, we limited the scope of the analysis to a maximum of 50 randomly selected SRAs for each of the 85 genera. These are listed in the supplementary file `sras.tab`.  The number of candidate reference genomes to be considered for each genus was capped at 200; the median number of genomes per genus was 9 (see Figure S3). The only genus with more than 200 reference genomes available was *Bordetella*. A maximum Mash distance was set to 0.1 [@ondov_mash_2016] between a preliminary assembly and a reference genome, as this was shown to be the maximum distance between the reference and sequenced isolate that riboSeed performs well with (see @waters_riboseed_2018 Figure 5). Run scripts are available in supplementary data; the reference genomes considered can be found  in Supplementary file `reference_genomes.tab`.


## Benchmarking Re-assembled 16S rRNA against Complete Genome and Draft 16S rRNA sequences {#benchmarking}

sraFind was used to identify which SRA accessions corresponded to complete NCBI genomes for the genera considered in this study. These were matched with SILVA sequences sourced from complete genomes (see supplementary data "Provenance of strains"). Pairwise alignments were generated between the riboSeed-assembled 16S rRNA sequences and the SILVA sequences using the Biostrings package [@pages_biostrings_2020] in "overlap" mode (a global alignment with free ends) with a simple scoring matrix  (matches=1, mismatches=0); the highest-scoring alignment for each given reference 16S rRNA was used to identify misassemblies relative to the complete genome's 16S rRNA sequence. Alignments shorter than 1400bp were rejected.

We identified the SILVA sequences sourced from draft genome assemblies (see supplementary data "Provenance of strains"), and were processed in the same manner as those from complete genomes


## Assessing Taxonomic Assignment
In order to determine whether this database augmentation approach would improve taxonomic assignment of ASVs, we used a DADA2-based pipeline was used to process each of the four datasets in Table \@ref(tab:dataset-tab).  The resulting sequence tables were combined, and we assigned taxonomy with the naive Bayes classifier implemented in DADA2. This classified sequences at the genus level, and DADA2's "assignSpecies" command was used to assign species-level taxonomy; we enabled the "allowMultiple" parameter to view ambiguities in the assignment. This analysis was used to compare assignment with SILVA 132 alone and assignment with SILVA 132 augmented with sequences generated by `r toolname`. All scripts can be found in supplementary materials.


# Results

##  Benchmarking Re-assembly Accuracy
Using sraFind, we identified which sequences in SILVA 132 originated from closed, complete genomes; those genomes with short-read SRAs were used to benchmark the accuracy of the 16S rRNA sequences re-assembled with the `r toolname` pipeline (as described in Methods section \@ref(benchmarking)) against the sequence in SILVA. Across the 85 genera we considered, 61 of these SRA/complete genome pairs were identified.


 The *de novo* assemblies had the lowest error rate but recovered the fewest alleles (Figure \@ref(fig:rdnamodes)). riboSeed's subassemblies had low error rates and successfully reconstructed the most 16S rDNAs, and as such are our recommendation for database augmentation as presented below.  Comparing the re-assembled 16S rRNA sequences to the 16S rRNA sequences from complete genomes shows a low error rate, 35.7% of sequences being perfect reconstructions across the entire full length 16S, and and 97% of sequences having fewer than 10 errors per Kbp  (Figure \@ref(fig:rdnamodes)B). This confirms that `r toolname`'s best-case accuracy yields perfect reconstructions of the full-length 16S rRNA region.  The V3-V4 region showed particularly encouraging results, recovering 197 error-free sequences.
  This suggests that sequences could be used to augment existing databases; the benefits and consequences of this are presented in the discussion. 
 
 
 The notable outliers are from the assembly of Corynebacterium sp. ATCC 6931 (SRR2889779/GCA_000755185.1);  it is unclear exactly why these re-assemblies were so poor, but @lupo_contamination_2021 recently flagged this assembly in their analysis as "ambiguous taxon", indicating a conflict between the reported taxonomy and that assigned by CheckM. 




The error rates for amplicon data in SILVA are difficult to determine; under optimal conditions Sanger sequencing has very low error rates [@shendure_nextgeneration_2008]. However when multiple sequences are inadvertently sequenced at the same time (i.e. multiple copies from a single organism), the trace will reflect the differences as short, imperfect, or overlapping peaks.  As the trace/quality data for amplicon sequences are not typically available, it is impossible to determine the accuracy of such sequences.

##  Comparing re-assembled 16S rRNA to draft 16S rRNA
```{r draftexamples, out.width="100%", fig.cap="Representative SNP alignments comparing 16S rRNA sequences from original draft assemblies to the re-assembled sequences. Alignments were generated with DECIPHER and all columns matching the consensus were removed; original alignment length and column numbers for each SNP are shown under the sequence names.  Names starting with an SRA accession such as ERR and SRR are the re-assembled sequences.  The following types of relationships occur:  all alleles recovered in original and re-assembled (A,B), sole original sequence misses a single (C) or multiple SNPs (D), disagreement between sole original and re-assembled alleles (E), original sequences appears to be amalgamation of alleles (FG), a deletion is present in original allele (H), re-assembly fails to reconstruct an allele (I)."}
knitr::include_graphics("figures/draftaln.png")

```

As repeated rDNA operons are difficult to resolve with short read sequences, draft genome assemblies can (and often do) contain a single assembled rDNA region with elevated read coverage. This can be problematic for genus or 16S rRNA classification as the 16S rRNA recovered may not just correspond to one of several 16S rRNA copies, but it can be a consensus "summary"/"collapsed" 16S rRNA resulting from imperfect assembly of the repeated region. We provide a few examples of such alignments in Figure  \@ref(fig:draftexamples) (see all alignments in supplementary information, in folder   `figures/draft_alignments/`).


In such cases, without the capacity to verify the regions with Sanger or long-read sequencing, determining which sequences are the missassemblies and which should be regarded as true is an impossible task.


##  Augmenting SILVA with results from `r toolname`
### Recovering Sequences from Re-assembly
<!-- (Figure \@ref(fig:sras)) -->
`r toolname` was used to build an extended database for the three mock datasets described in the DADA2 paper and a real-world dataset from the Endobiota study. From the 85 genera considered, `r toolname` processed 2387 SRAs, and recovered 16S rRNA sequences from 1392 SRAs (Figure \@ref(fig:reassembly). The average execution time for each SRA was approximately 23 minutes. Several factors can contribute to failing to recover 16S rRNA sequences from a given SRA, and among these are a too-distant reference genome, low rDNA flanking diversity, short read length, or high read error rates. In total, we recovered 5647 16S rRNA sequences, of which 2823 were unique. Likely due to variations in the trimming of the sequences, none of the 2823 exactly matched existing matched existing SILVA sequences, but that is 


### Recovery of unique sequences
In this pilot study, we assessed the increase in unique sequences gained by augmenting SILVA with only the 85 genera found across these four datasets. For thoroughly-sequenced genera such as *Escherichia*, *Pseudomonas*, or *Bacillus*, the increases in unique sequences  are small. However, other taxa showed marked increases in the genus-level 16S rRNA sequence diversity; for example, there was a 75% increase in the number of *Aerococcus* sequences.  (Figure \@ref(fig:diversity)).


```{r reassembly, echo=FALSE, fig.cap="Summary of Focus16 outcomes for reassembling SRA datasets. Green circles indicate SRAs that yielded 16S rRNA sequences, while blue diamonds indicate SRAs failing to yield re-assembled 16S rRNA sequences. Pink triangles show SRAs that were rejected due to limitations in the diversity of available reference genomes, and inverted green triangles show SRAs rejected due to read length, insufficient coverage, poor read quality, etc. A few errors occurred, usually when the SRA metadata conflicted with the associated sequencing data and caused download errors or errors from reads with incorrect pairing.  In these cases, the datasets were discarded.", out.width = '80%'}
knitr::include_graphics("figures/2-genera.png")
```

## Assessing Taxonomic Assignment
Having shown that `r toolname` can recover novel 16S rRNA alleles from re-assembly of draft genomes, we quantified the effect of including these sequences when assigning taxonomy to ASVs.
DADA2 was used to identify ASVs from four datasets (Table 1), resulting in a total of 4098 ASVs (109 sequences from the "HMP" dataset, 26 from the "Extreme" dataset, 94 from the "balanced" dataset, and the remaining 3869  from the EndoBiota study). We then compared the taxonomic results of classification using the SILVA 132 database with or without novel 16S rRNA alleles identified with `r toolname`.

The taxonomic assignments of 20 strains differed when using the augmented focus16 database (see Table 2, or Supplementary file `STABLE_different_assignment.tab` for the ASV sequences). Changes could happen in three ways: a previously unclassified ASV gained classification, a previously-assigned ASV gained more species- (or genus-) level details, or a previously-classified ASV was given a different classification. In our dataset, three previously-unclassified strains gained annotations (Table 2 rows 5, 10, and 11) . The remaining 17 had more detail added to the genus or species level; typically, assignment with SILVA resulted in a single species classification, but the augmented database revealed the ASV to be ambiguous either at the species (Table 2 rows 4,6,8,9,12,13,14 ) or genus (Table 2 rows 1,2,7, 15-20) level. 

If this pilot study is perfectly representative, users could expect an improved taxonomic assignment for about 0.5% of ASVs; two factors must be considered before extrapolating that value beyond this study.  First, we were limited in the number of SRAs per genus that could be processed (Supplementary Figure 1 shows the number of SRAs per genus). Second, and perhaps more importantly, the majority of genera in this study are associated with human microbiomes, an area which has already seen an extensive amount of focus in terms of amplicon sequencing, whole-genome sequencing, and genome completion efforts. Other environments have not had this benefit, and may benefit even more from this approach.

```{r diversity, echo=FALSE, fig.cap=paste0("Percentage increase in unique sequences (and number of added unique sequences), by genus. The addition of the sequences recovered by ", toolname, " increases the number of unique sequences for the given taxa; the greatest increases are found in \"under-sequenced\" taxa and/or taxa with less well-conserved rRNA sequences."), out.width = '80%'}
knitr::include_graphics("figures/3-perc_unique_genus.png")

```

```{r hits-tab, size="tiny"}
knitr::kable(
  booktabs=TRUE,
  caption="Of the 4098 ASVs aggregated across the four datasets considered, the augmented database modified  taxonomic assignment of 20 sequences. Cells reading 'NA NA' indicate that no genus or species could be assigned. Cells reading NA speciesA/speciesB indicate that the lowest common ancestor for a sequence could not be determined at the genus level; in such cases DADA2 gives the designation NA rather than listing possible genera as is done at species level.", col.names =  c("", "SILVA", paste("SILVA + ", toolname)),
  read.csv("../STABLE_different_assignment.tab", sep="\t") %>%
    mutate(set=row_number()) %>%
    select(set, silva, focuDB) %>%
    transform(
      silva =gsub("\\/", "\\/ ", silva),
      focuDB =gsub("\\/", "\\/ ", focuDB))
) %>% column_spec(1, width = "2em")   %>% column_spec(2, width = "20em")  %>% column_spec(3, width = "20em")  %>% kable_styling(latex_options = "striped")
```


# Discussion
`r toolname` orchestrates the re-assembly of whole-genome sequencing datasets in SRA to recover 16S rRNA alleles that may be missing from the existing reference databases. Using riboSeed, `r toolname` re-assembles draft genomes that currently contribute a single (often collapsed) 16S rRNA to resolve distinct instances of the 16S rRNA operon. We show that this increases the sequence diversity (number of unique sequences) of the 16S rRNA databases, and that the increased diversity results in measurable improvements to taxonomic assignment.



`r toolname` improved fine-scale taxonomic assignment in two ways: first, by assigning previously unclassified sequences and second, by correcting spurious species-level assignment for ASVs that could come from two or more species. While at face value this appears to reduce the precision of taxonomic assignment, it reveals cases where species-level assignment was inappropriate.  Based on the improved taxonomic assignment in this pilot study of 85 genera, we believe a wide-scale application of `r toolname` could benefit the community.

A natural concern about the approach presented here is the danger of "poisoning" the database with sequences that may or may not be 100% accurate. This is a valid concern, but one we believe to be outweighed by the potential of offsetting the known problems with existing 16S rRNA databases. Those 16S rRNA sequences in SILVA originating from amplicons lack the taxonomic certainty that comes with whole-genome sequencing. Sequences from draft whole-genome assemblies have known issues with rDNA missassembly; in the best case, only one accurate 16S rRNA is represented; in the worst case, the one assembled 16S rRNA sequence is an amalgamation of an unknown number of distinct sequence variants. Until long-read or whole-metagenome sequencing efforts fully supplant short-read amplicon sequencing, these issues must be considered when attempting community analysis via 16S rRNA sequencing. A conservative approach to utilizing sequences recovered with the methodology presented here may be to incorporate a measure of taxonomic assignment confidence, where references sequences originating from long reads, amplicons, draft assemblies, and focus16 reassemblies could be weighted appropriately. 

<!-- ((98329*(23/60)) /24)/150 -->
However, there are three main limitations facing the large-scale application of Focus16: the first is the bandwidth, computational power, memory, and storage required to re-assemble the 98,329 SRAs (as of October 2019) that were used to generate draft genomes. Given sufficient storage and unfettered access to a medium/large computing cluster such as one supporting a university or research institute (say 150 compute nodes), the task could be completed within two weeks^[This estimation is based on an rough average processing time of 23 minutes per run, but this is highly dependent on download speed, read/write speed, genome size, and sequencing depth.], and the task could be accomplished in even less time with a sufficiently large cloud computing budget; however with modest hardware (8 cores, 20GB RAM), this would take about 4 years in “--fast” mode. These estimates are ignoring the 112,695 draft genomes for which no reads were ever released, which leads to the second limitation: data availability. Releasing draft genomes without the reads used to generate them hampers efforts such as this one to expand beyond the purpose of the original study.

The third limitation of this approach is the availability of high-quality closed genomes to use as references. With the increased adoption of long read technologies, we envisage that this limitation will decrease with time; re-running the pipeline as new, complete reference genomes are generated will allow for ongoing improvements to the databases. Eventually, a point will come when `r toolname` will no longer be needed as all candidate SRAs have been sufficiently utilized.

Further limitations exist within `r toolname` and within riboSeed. The success of riboSeed's *de fere novo* assembly hinges on the similarity of the reference to the sequenced isolate, the differentiating power of the rDNA flanking regions, read length, and other factors. This is one reason why not all SRAs yielded perfect rDNAs. Additionally, riboSeed does not currently support mate-paired libraries; these are much less widely used than the typical single-end or paired-end libraries used in short-read sequencing.

Despite these limitations, we have shown that `r toolname` can contribute towards better molecular ecology analysis;  augmenting SILVA with the sequences re-assembled from the 85 genera considered here led to a small increase in the number of unique sequences in the database.  Using the augmented database for taxonomic assignment revealed some limitations of low-level taxonomic assignment, and led to the classification of additional ASVs.  We invite the community to consider augmenting existing databases (such as NCBI's 16S RefSeq Microbial database, SILVA, RDP, GTDB, and GreenGenes) according to the approach outlined here.



# Competing interests
The authors declare that they have no competing interests.

# Funding
This work was financially supported by Science Foundation Ireland (Awards 14/IA/2371 and 16/RC/3889) and through  a joint studentship between The James Hutton Institute and the National University of Ireland, Galway.

# Acknowledgements
Many thanks to Christopher Quince for the helpful conversations on the topic.

# Data Accessibility
The code for Focus16 can be found at <https://github.com/FEMLab/focus16>; the code for all the analyses presented in this work can be found at <https://github.com/FEMLab/focus16_manuscript>. All data used is archived at Zenodo accession 10.5281/zenodo.3956433.

# Author Contributions
Author contributions according to the CRediT taxonomy [@allen_how_2019] are listed alphabetically as follows:
Conceptualization: LP, NW ; Methodology: FA, BN, LP, NW; Software and Data Curation: BN, NW; Validation: BN, LP, NW; Formal analysis: BN, LP, NW; Investigation: BN, NW; Resources; FA, VOF, LP; Writing - Original Draft: BN, NW; Writing - Review & Editing: FA, FB, VOF, AH,  BN, LP, NW; Visualization: BN, LP, NW; Supervision: FA, FB, VOF, AH, LP, NW; Project administration	FA, VOF, NW; Funding acquisition: FA, FB, VOF, LP, NW.


# References
